{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2177a486",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy\n",
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "from statsmodels.stats.multitest import fdrcorrection,multipletests\n",
    "\n",
    "# Set the default font to Arial\n",
    "import matplotlib\n",
    "matplotlib.rcParams['font.family'] = 'Arial'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c8cd7a9",
   "metadata": {},
   "source": [
    "# Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3d08171f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The filtering based on hotspots.\n",
    "filter_version = 'including'  # 'including', 'excluding', 'only'\n",
    "list_filter_versions = ['including', 'excluding', 'only']\n",
    "\n",
    "PATH = '../../results/chrM_2024/'\n",
    "\n",
    "INDEL_PREFIX = {\n",
    "        'including':'indels/indels.denovo_', \n",
    "        'excluding':'indels/indels.exclHotspots.denovo_', \n",
    "        'only':'indels/indels.onlyHotspots.denovo_' \n",
    "    }[filter_version]\n",
    "\n",
    "FREQS_PREFIX = {\n",
    "        'including':'tables/freqs.denovo_indels.', \n",
    "        'excluding':'tables/freqs.exclHotspots.', \n",
    "        'only':'tables/freqs.onlyHotspots.' \n",
    "    }[filter_version]\n",
    "\n",
    "\n",
    "# The 3 species.\n",
    "list_species = ['mouse','macaque','human']\n",
    "# The alignment chosen.\n",
    "alignment = 'chrM'\n",
    "# Dictionary of tissues per species.\n",
    "dict_tissues =  {\n",
    "                    'mouse':['Oo', 'M', 'Br'], \n",
    "                    'macaque':['Oo', 'M', 'Li'],\n",
    "                    'human':['Oo', 'Sa', 'Bl'],\n",
    "                }\n",
    "dict_tissue_names = {'Oo':'Oocytes','M':'Skeletal muscle','Br':'Brain',\n",
    "                'Li':'Liver','Sa':'Saliva','Bl':'Blood'}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bcecf8d",
   "metadata": {},
   "source": [
    "# Wrangle CDS annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73705574",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "\n",
    "dict_files = {'macaque':'sequence.rhesus_macaque.NC_005943.gb', \n",
    "\t\t\t  'mouse':'sequence.mouse.NC_005089.gb', \n",
    "\t\t\t  'human':'sequence.human.NC_012920.gb'}\n",
    "\n",
    "for species in dict_files:\n",
    "\n",
    "\t# Parse the GenBank file using SeqIO\n",
    "\tgb_file = dict_files[species]\n",
    "\trecord = SeqIO.read(gb_file, 'genbank')\n",
    "\n",
    "\t# Extract key information\n",
    "\tprint(f\"Sequence ID: {record.id}\")\n",
    "\tprint(f\"Description: {record.description}\")\n",
    "\tprint(f\"Sequence length: {len(record.seq)}\")\n",
    "\tprint(f\"Number of features: {len(record.features)}\")\n",
    "\n",
    "\t# Display features\n",
    "\t# for feature in record.features:\n",
    "\t\t# print(f\"Feature type: {feature.type}, Location: {feature.location}\")\n",
    "\n",
    "\t# Filter CDS features\n",
    "\tcds_features = [feature for feature in record.features if feature.type == 'CDS']\n",
    "\tprint(f\"\\nNumber of CDS features: {len(cds_features)}\")\n",
    "\n",
    "\t# Display CDS features\n",
    "\tfor cds in cds_features:\n",
    "\t\tprint(f\"CDS Location: {cds.location}, Qualifiers: {cds.qualifiers}\")\n",
    "\n",
    "\t\t# Convert CDS features to a pandas DataFrame\n",
    "\t\tcds_data = []\n",
    "\t\tfor cds in cds_features:\n",
    "\t\t\tcds_data.append({\n",
    "\t\t\t\t'Location': str(cds.location),\n",
    "\t\t\t\t'Start': cds.location.start,\n",
    "\t\t\t\t'End': cds.location.end,\n",
    "\t\t\t\t'Strand': cds.location.strand,\n",
    "\t\t\t\t'Qualifiers': cds.qualifiers\n",
    "\t\t\t})\n",
    "\n",
    "\t\tcds_df = pd.DataFrame(cds_data)\n",
    "\t\tcds_df\n",
    "\n",
    "\t# Extract coordinates as a list in BED format\n",
    "\tcoordinates = cds_df[['Start', 'End', 'Strand']].values.tolist()\n",
    "\t# Use 0-based start positions\n",
    "\tcoordinates = [[c[0] - 1, c[1], c[2]] for c in coordinates]\n",
    "\tcoordinates\n",
    "\n",
    "\t# Write coordinates to BED file\n",
    "\tout_file = gb_file.replace('.gb', '.cds.bed')\n",
    "\twith open(out_file, 'w') as bed_file:\n",
    "\t\tfor i, coord in enumerate(coordinates):\n",
    "\t\t\tchrom = record.id\n",
    "\t\t\tstart = coord[0]\n",
    "\t\t\tend = coord[1]\n",
    "\t\t\tstrand = '+' if coord[2] == 1 else '-'\n",
    "\t\t\tbed_file.write(f\"{chrom}\\t{start}\\t{end}\\t.\\t.\\t{strand}\\n\")\n",
    "\n",
    "\tprint(f\"\\nBED file created: {out_file}\\n\")\n",
    "\n",
    "\twith open(out_file, 'r') as f:\n",
    "\t\tprint(f.read())\n",
    "\n",
    "# Merge overlapping CDS regions using bedtools.\n",
    "!bedtools merge -i sequence.human.NC_012920.cds.bed > sequence.human.NC_012920.cds.merged.bed\n",
    "!bedtools merge -i sequence.mouse.NC_005089.cds.bed > sequence.mouse.NC_005089.cds.merged.bed\n",
    "!bedtools merge -i sequence.rhesus_macaque.NC_005943.cds.bed > sequence.rhesus_macaque.NC_005943.cds.merged.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16131987",
   "metadata": {},
   "source": [
    "# Import CDS annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "79e01a5b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>macaque</td>\n",
       "      <td>3257</td>\n",
       "      <td>4213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>macaque</td>\n",
       "      <td>4419</td>\n",
       "      <td>5462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>macaque</td>\n",
       "      <td>5848</td>\n",
       "      <td>7418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>macaque</td>\n",
       "      <td>7530</td>\n",
       "      <td>8215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>macaque</td>\n",
       "      <td>8355</td>\n",
       "      <td>9981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>macaque</td>\n",
       "      <td>10048</td>\n",
       "      <td>10395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>macaque</td>\n",
       "      <td>10459</td>\n",
       "      <td>12128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>macaque</td>\n",
       "      <td>12326</td>\n",
       "      <td>14667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>macaque</td>\n",
       "      <td>14739</td>\n",
       "      <td>15881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mouse</td>\n",
       "      <td>2749</td>\n",
       "      <td>3707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mouse</td>\n",
       "      <td>3912</td>\n",
       "      <td>4951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mouse</td>\n",
       "      <td>5326</td>\n",
       "      <td>6872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mouse</td>\n",
       "      <td>7011</td>\n",
       "      <td>7696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mouse</td>\n",
       "      <td>7764</td>\n",
       "      <td>9390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>mouse</td>\n",
       "      <td>9457</td>\n",
       "      <td>9806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>mouse</td>\n",
       "      <td>9876</td>\n",
       "      <td>11545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>mouse</td>\n",
       "      <td>11741</td>\n",
       "      <td>14071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>mouse</td>\n",
       "      <td>14144</td>\n",
       "      <td>15289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>human</td>\n",
       "      <td>3305</td>\n",
       "      <td>4262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>human</td>\n",
       "      <td>4468</td>\n",
       "      <td>5511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human</td>\n",
       "      <td>5902</td>\n",
       "      <td>7445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>human</td>\n",
       "      <td>7584</td>\n",
       "      <td>8269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human</td>\n",
       "      <td>8364</td>\n",
       "      <td>9990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>human</td>\n",
       "      <td>10057</td>\n",
       "      <td>10404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>human</td>\n",
       "      <td>10468</td>\n",
       "      <td>12137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>human</td>\n",
       "      <td>12335</td>\n",
       "      <td>14673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>human</td>\n",
       "      <td>14745</td>\n",
       "      <td>15887</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   species  start    end\n",
       "0  macaque   3257   4213\n",
       "1  macaque   4419   5462\n",
       "2  macaque   5848   7418\n",
       "3  macaque   7530   8215\n",
       "4  macaque   8355   9981\n",
       "5  macaque  10048  10395\n",
       "6  macaque  10459  12128\n",
       "7  macaque  12326  14667\n",
       "8  macaque  14739  15881\n",
       "0    mouse   2749   3707\n",
       "1    mouse   3912   4951\n",
       "2    mouse   5326   6872\n",
       "3    mouse   7011   7696\n",
       "4    mouse   7764   9390\n",
       "5    mouse   9457   9806\n",
       "6    mouse   9876  11545\n",
       "7    mouse  11741  14071\n",
       "8    mouse  14144  15289\n",
       "0    human   3305   4262\n",
       "1    human   4468   5511\n",
       "2    human   5902   7445\n",
       "3    human   7584   8269\n",
       "4    human   8364   9990\n",
       "5    human  10057  10404\n",
       "6    human  10468  12137\n",
       "7    human  12335  14673\n",
       "8    human  14745  15887"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def get_cds():\n",
    "\tdict_files = {'macaque':'sequence.rhesus_macaque.NC_005943.cds.merged.bed', \n",
    "\t\t\t  \t\t\t'mouse':'sequence.mouse.NC_005089.cds.9821insA.merged.bed', \n",
    "\t\t\t  \t\t\t'human':'sequence.human.NC_012920.cds.merged.bed'}\n",
    "\tdict_genome_sizes = {'macaque':16564, 'mouse':16300, 'human':16569}\n",
    "\tfor species in dict_files.keys():\n",
    "\t\tcds_file = dict_files[species]\n",
    "\t\tdf_cds = pd.read_csv(cds_file, sep='\\t', header=None, names=['chrom','start','end'])\n",
    "\t\tdf_cds['species'] = species\n",
    "\tdf_merged = pd.concat( [pd.read_csv(dict_files[sp], sep='\\t', header=None, names=['chrom','start','end']).assign(species=sp) for sp in dict_files.keys()] )\n",
    "\tdf_merged['genome_size'] = df_merged['species'].map(dict_genome_sizes)\n",
    "\tdf_merged['cds_length'] = df_merged['end'] - df_merged['start']\n",
    "\tdf_cds = df_merged[['species','start','end']]\n",
    "\t\n",
    "\treturn df_cds\n",
    "\n",
    "df_cds = get_cds()\n",
    "df_cds\t\t"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e014aa4",
   "metadata": {},
   "source": [
    "# Import indels for 3 species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c258713c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>INFO</th>\n",
       "      <th>SRR_ID</th>\n",
       "      <th>value</th>\n",
       "      <th>Species</th>\n",
       "      <th>AC</th>\n",
       "      <th>AF</th>\n",
       "      <th>...</th>\n",
       "      <th>Tissue_type</th>\n",
       "      <th>Classification</th>\n",
       "      <th>Mutation_ID</th>\n",
       "      <th>Pedigree_shared</th>\n",
       "      <th>Filter_ID</th>\n",
       "      <th>Region</th>\n",
       "      <th>Region_size</th>\n",
       "      <th>Indel_length</th>\n",
       "      <th>Indel_type</th>\n",
       "      <th>is_CDS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrM</td>\n",
       "      <td>12343</td>\n",
       "      <td>AT</td>\n",
       "      <td>A</td>\n",
       "      <td>SB=2.42169;AF=0.00495049;AC=0</td>\n",
       "      <td>SRR10068699</td>\n",
       "      <td>.:1:0.00495049:2.42169</td>\n",
       "      <td>mouse</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.004950</td>\n",
       "      <td>...</td>\n",
       "      <td>germline</td>\n",
       "      <td>Germline de novo</td>\n",
       "      <td>12343_AT_A</td>\n",
       "      <td>Unique to Individual</td>\n",
       "      <td>G133p4_12343</td>\n",
       "      <td>non-D-loop</td>\n",
       "      <td>15423</td>\n",
       "      <td>1</td>\n",
       "      <td>deletion</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrM</td>\n",
       "      <td>13053</td>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "      <td>SB=2.11058;AF=0.00329761;AC=0</td>\n",
       "      <td>SRR10068701</td>\n",
       "      <td>.:2:0.00115607:2.64192</td>\n",
       "      <td>mouse</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.001156</td>\n",
       "      <td>...</td>\n",
       "      <td>germline</td>\n",
       "      <td>Germline de novo</td>\n",
       "      <td>13053_T_TC</td>\n",
       "      <td>Between pedigrees</td>\n",
       "      <td>G133p4_13053</td>\n",
       "      <td>non-D-loop</td>\n",
       "      <td>15423</td>\n",
       "      <td>1</td>\n",
       "      <td>insertion</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrM</td>\n",
       "      <td>14806</td>\n",
       "      <td>CACCCCTACTATACAATCAAAGATATCCTAGGTATCCTAATCATAT...</td>\n",
       "      <td>C</td>\n",
       "      <td>SB=2.57858;AF=0.000629327;AC=0</td>\n",
       "      <td>SRR10068701</td>\n",
       "      <td>.:1:0.000629327:2.57858</td>\n",
       "      <td>mouse</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000629</td>\n",
       "      <td>...</td>\n",
       "      <td>germline</td>\n",
       "      <td>Germline de novo</td>\n",
       "      <td>14806_CACCCCTACTATACAATCAAAGATATCCTAGGTATCCTAA...</td>\n",
       "      <td>Unique to Individual</td>\n",
       "      <td>G133p4_14806</td>\n",
       "      <td>non-D-loop</td>\n",
       "      <td>15423</td>\n",
       "      <td>58</td>\n",
       "      <td>deletion</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrM</td>\n",
       "      <td>14640</td>\n",
       "      <td>A</td>\n",
       "      <td>AG</td>\n",
       "      <td>SB=1.61939;AF=0.00191388;AC=0</td>\n",
       "      <td>SRR10068702</td>\n",
       "      <td>.:2:0.00191388:1.61939</td>\n",
       "      <td>mouse</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.001914</td>\n",
       "      <td>...</td>\n",
       "      <td>somatic</td>\n",
       "      <td>Somatic de novo</td>\n",
       "      <td>14640_A_AG</td>\n",
       "      <td>Unique to Individual</td>\n",
       "      <td>G133p4_14640</td>\n",
       "      <td>non-D-loop</td>\n",
       "      <td>15423</td>\n",
       "      <td>1</td>\n",
       "      <td>insertion</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrM</td>\n",
       "      <td>13053</td>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "      <td>SB=2.11058;AF=0.00329761;AC=0</td>\n",
       "      <td>SRR10068708</td>\n",
       "      <td>.:2:0.00754717:2.07087</td>\n",
       "      <td>mouse</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.007547</td>\n",
       "      <td>...</td>\n",
       "      <td>somatic</td>\n",
       "      <td>Somatic de novo</td>\n",
       "      <td>13053_T_TC</td>\n",
       "      <td>Between pedigrees</td>\n",
       "      <td>G131p1_13053</td>\n",
       "      <td>non-D-loop</td>\n",
       "      <td>15423</td>\n",
       "      <td>1</td>\n",
       "      <td>insertion</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11030</td>\n",
       "      <td>GA</td>\n",
       "      <td>G</td>\n",
       "      <td>SB=1.51818;AF=0.00198807;AC=0</td>\n",
       "      <td>hs023_Sa</td>\n",
       "      <td>.:1:0.00198807:1.41855</td>\n",
       "      <td>human</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.001988</td>\n",
       "      <td>...</td>\n",
       "      <td>somatic</td>\n",
       "      <td>Somatic de novo</td>\n",
       "      <td>11030_GA_G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hs023_11030</td>\n",
       "      <td>non-D-loop</td>\n",
       "      <td>15447</td>\n",
       "      <td>1</td>\n",
       "      <td>deletion</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11865</td>\n",
       "      <td>A</td>\n",
       "      <td>AC</td>\n",
       "      <td>SB=1.80374;AF=0.00170358;AC=0</td>\n",
       "      <td>hs023_Sa</td>\n",
       "      <td>.:1:0.00170358:1.80339</td>\n",
       "      <td>human</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.001704</td>\n",
       "      <td>...</td>\n",
       "      <td>somatic</td>\n",
       "      <td>Somatic de novo</td>\n",
       "      <td>11865_A_AC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hs023_11865</td>\n",
       "      <td>non-D-loop</td>\n",
       "      <td>15447</td>\n",
       "      <td>1</td>\n",
       "      <td>insertion</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>chrM</td>\n",
       "      <td>12383</td>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "      <td>SB=1.91172;AF=0.0018315;AC=0</td>\n",
       "      <td>hs023_Sa</td>\n",
       "      <td>.:1:0.0018315:1.91482</td>\n",
       "      <td>human</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.001832</td>\n",
       "      <td>...</td>\n",
       "      <td>somatic</td>\n",
       "      <td>Somatic de novo</td>\n",
       "      <td>12383_T_TC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hs023_12383</td>\n",
       "      <td>non-D-loop</td>\n",
       "      <td>15447</td>\n",
       "      <td>1</td>\n",
       "      <td>insertion</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>chrM</td>\n",
       "      <td>13131</td>\n",
       "      <td>CT</td>\n",
       "      <td>C</td>\n",
       "      <td>SB=1.79468;AF=0.00168067;AC=0</td>\n",
       "      <td>hs023_Sa</td>\n",
       "      <td>.:1:0.00168067:1.79468</td>\n",
       "      <td>human</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.001681</td>\n",
       "      <td>...</td>\n",
       "      <td>somatic</td>\n",
       "      <td>Somatic de novo</td>\n",
       "      <td>13131_CT_C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hs023_13131</td>\n",
       "      <td>non-D-loop</td>\n",
       "      <td>15447</td>\n",
       "      <td>1</td>\n",
       "      <td>deletion</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>chrM</td>\n",
       "      <td>14073</td>\n",
       "      <td>CA</td>\n",
       "      <td>C</td>\n",
       "      <td>SB=1.60975;AF=0.00185185;AC=0</td>\n",
       "      <td>hs023_Sa</td>\n",
       "      <td>.:1:0.00185185:1.60975</td>\n",
       "      <td>human</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.001852</td>\n",
       "      <td>...</td>\n",
       "      <td>somatic</td>\n",
       "      <td>Somatic de novo</td>\n",
       "      <td>14073_CA_C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hs023_14073</td>\n",
       "      <td>non-D-loop</td>\n",
       "      <td>15447</td>\n",
       "      <td>1</td>\n",
       "      <td>deletion</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1630 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    CHROM    POS                                                REF ALT  \\\n",
       "0    chrM  12343                                                 AT   A   \n",
       "1    chrM  13053                                                  T  TC   \n",
       "2    chrM  14806  CACCCCTACTATACAATCAAAGATATCCTAGGTATCCTAATCATAT...   C   \n",
       "3    chrM  14640                                                  A  AG   \n",
       "5    chrM  13053                                                  T  TC   \n",
       "..    ...    ...                                                ...  ..   \n",
       "130  chrM  11030                                                 GA   G   \n",
       "131  chrM  11865                                                  A  AC   \n",
       "132  chrM  12383                                                  T  TC   \n",
       "133  chrM  13131                                                 CT   C   \n",
       "134  chrM  14073                                                 CA   C   \n",
       "\n",
       "                               INFO       SRR_ID                    value  \\\n",
       "0     SB=2.42169;AF=0.00495049;AC=0  SRR10068699   .:1:0.00495049:2.42169   \n",
       "1     SB=2.11058;AF=0.00329761;AC=0  SRR10068701   .:2:0.00115607:2.64192   \n",
       "2    SB=2.57858;AF=0.000629327;AC=0  SRR10068701  .:1:0.000629327:2.57858   \n",
       "3     SB=1.61939;AF=0.00191388;AC=0  SRR10068702   .:2:0.00191388:1.61939   \n",
       "5     SB=2.11058;AF=0.00329761;AC=0  SRR10068708   .:2:0.00754717:2.07087   \n",
       "..                              ...          ...                      ...   \n",
       "130   SB=1.51818;AF=0.00198807;AC=0     hs023_Sa   .:1:0.00198807:1.41855   \n",
       "131   SB=1.80374;AF=0.00170358;AC=0     hs023_Sa   .:1:0.00170358:1.80339   \n",
       "132    SB=1.91172;AF=0.0018315;AC=0     hs023_Sa    .:1:0.0018315:1.91482   \n",
       "133   SB=1.79468;AF=0.00168067;AC=0     hs023_Sa   .:1:0.00168067:1.79468   \n",
       "134   SB=1.60975;AF=0.00185185;AC=0     hs023_Sa   .:1:0.00185185:1.60975   \n",
       "\n",
       "    Species   AC        AF  ...  Tissue_type    Classification  \\\n",
       "0     mouse  1.0  0.004950  ...     germline  Germline de novo   \n",
       "1     mouse  2.0  0.001156  ...     germline  Germline de novo   \n",
       "2     mouse  1.0  0.000629  ...     germline  Germline de novo   \n",
       "3     mouse  2.0  0.001914  ...      somatic   Somatic de novo   \n",
       "5     mouse  2.0  0.007547  ...      somatic   Somatic de novo   \n",
       "..      ...  ...       ...  ...          ...               ...   \n",
       "130   human  1.0  0.001988  ...      somatic   Somatic de novo   \n",
       "131   human  1.0  0.001704  ...      somatic   Somatic de novo   \n",
       "132   human  1.0  0.001832  ...      somatic   Somatic de novo   \n",
       "133   human  1.0  0.001681  ...      somatic   Somatic de novo   \n",
       "134   human  1.0  0.001852  ...      somatic   Somatic de novo   \n",
       "\n",
       "                                           Mutation_ID       Pedigree_shared  \\\n",
       "0                                           12343_AT_A  Unique to Individual   \n",
       "1                                           13053_T_TC     Between pedigrees   \n",
       "2    14806_CACCCCTACTATACAATCAAAGATATCCTAGGTATCCTAA...  Unique to Individual   \n",
       "3                                           14640_A_AG  Unique to Individual   \n",
       "5                                           13053_T_TC     Between pedigrees   \n",
       "..                                                 ...                   ...   \n",
       "130                                         11030_GA_G                   NaN   \n",
       "131                                         11865_A_AC                   NaN   \n",
       "132                                         12383_T_TC                   NaN   \n",
       "133                                         13131_CT_C                   NaN   \n",
       "134                                         14073_CA_C                   NaN   \n",
       "\n",
       "        Filter_ID      Region Region_size Indel_length Indel_type is_CDS  \n",
       "0    G133p4_12343  non-D-loop       15423            1   deletion   True  \n",
       "1    G133p4_13053  non-D-loop       15423            1  insertion   True  \n",
       "2    G133p4_14806  non-D-loop       15423           58   deletion   True  \n",
       "3    G133p4_14640  non-D-loop       15423            1  insertion   True  \n",
       "5    G131p1_13053  non-D-loop       15423            1  insertion   True  \n",
       "..            ...         ...         ...          ...        ...    ...  \n",
       "130   hs023_11030  non-D-loop       15447            1   deletion   True  \n",
       "131   hs023_11865  non-D-loop       15447            1  insertion   True  \n",
       "132   hs023_12383  non-D-loop       15447            1  insertion   True  \n",
       "133   hs023_13131  non-D-loop       15447            1   deletion   True  \n",
       "134   hs023_14073  non-D-loop       15447            1   deletion   True  \n",
       "\n",
       "[1630 rows x 35 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_indels(filter_version):\n",
    "    # Combine indels for the three species.\n",
    "    INDEL_PREFIX = {\n",
    "        'including':'indels/indels.denovo_', \n",
    "        'excluding':'indels/indels.exclHotspots.denovo_', \n",
    "        'only':'indels/indels.onlyHotspots.denovo_' \n",
    "    }\n",
    "    list_df = []\n",
    "    for species in list_species:\n",
    "        df = pd.read_table( f'{PATH+INDEL_PREFIX[filter_version]}{alignment}_{species}.tab' )\n",
    "        list_df.append( df )\n",
    "    df_indels = pd.concat(list_df)\n",
    "    df_indels['Indel_length'] = abs(df_indels['REF'].str.len() - df_indels['ALT'].str.len())\n",
    "    df_indels['Indel_type'] = np.where( df_indels['REF'].str.len() < df_indels['ALT'].str.len(), 'insertion', 'deletion' )  \n",
    "    \n",
    "    # Annotate as CDS or non-CDS.\n",
    "    df_cds = get_cds()\n",
    "    def is_in_cds(row):\n",
    "        species = row['Species']\n",
    "        pos = row['POS']\n",
    "        df_cds_sp = df_cds[ df_cds['species'] == species ]\n",
    "        for _, cds_row in df_cds_sp.iterrows():\n",
    "            if cds_row['start'] <= pos < cds_row['end']:\n",
    "                return True\n",
    "        return False\n",
    "    df_indels['is_CDS'] = df_indels.apply(is_in_cds, axis=1)\n",
    "    \n",
    "    # Only CDS indels\n",
    "    df_indels = df_indels[ df_indels['is_CDS'] ]\n",
    "\n",
    "    return df_indels\n",
    "\n",
    "df_indels = get_indels('including')\n",
    "df_indels"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d273ff7f",
   "metadata": {},
   "source": [
    "# Table indel lengths by species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b4bf2925",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Indel_type</th>\n",
       "      <th>Indel_length</th>\n",
       "      <th>mult3</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mouse</td>\n",
       "      <td>deletion</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mouse</td>\n",
       "      <td>deletion</td>\n",
       "      <td>2</td>\n",
       "      <td>.</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mouse</td>\n",
       "      <td>deletion</td>\n",
       "      <td>3</td>\n",
       "      <td>mult3</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mouse</td>\n",
       "      <td>deletion</td>\n",
       "      <td>4</td>\n",
       "      <td>.</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mouse</td>\n",
       "      <td>deletion</td>\n",
       "      <td>5</td>\n",
       "      <td>.</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>human</td>\n",
       "      <td>insertion</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>human</td>\n",
       "      <td>insertion</td>\n",
       "      <td>3</td>\n",
       "      <td>mult3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>human</td>\n",
       "      <td>insertion</td>\n",
       "      <td>6</td>\n",
       "      <td>mult3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>human</td>\n",
       "      <td>insertion</td>\n",
       "      <td>7</td>\n",
       "      <td>.</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>human</td>\n",
       "      <td>insertion</td>\n",
       "      <td>26</td>\n",
       "      <td>.</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>99 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Species Indel_type  Indel_length  mult3  Count\n",
       "0    mouse   deletion             1      .    133\n",
       "1    mouse   deletion             2      .     18\n",
       "2    mouse   deletion             3  mult3     15\n",
       "3    mouse   deletion             4      .      1\n",
       "4    mouse   deletion             5      .      5\n",
       "..     ...        ...           ...    ...    ...\n",
       "94   human  insertion             1      .     23\n",
       "95   human  insertion             3  mult3      1\n",
       "96   human  insertion             6  mult3      1\n",
       "97   human  insertion             7      .      1\n",
       "98   human  insertion            26      .      1\n",
       "\n",
       "[99 rows x 5 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table_lengths = pd.DataFrame(df_indels.groupby(['Species','Indel_type','Indel_length']).size()).reset_index().rename(columns={0:'Count'})\n",
    "# enforce desired species order (mouse, macaque, human) then sort\n",
    "table_lengths['Species'] = pd.Categorical(table_lengths['Species'], categories=list_species, ordered=True)\n",
    "table_lengths = table_lengths.sort_values(['Species','Indel_type','Indel_length']).reset_index(drop=True)\n",
    "table_lengths['mult3'] = (table_lengths['Indel_length'] % 3 == 0)\n",
    "table_lengths['mult3'] = table_lengths['mult3'].map({True: 'mult3', False: '.'}).astype('category')\n",
    "table_lengths = table_lengths[['Species', 'Indel_type', 'Indel_length', 'mult3', 'Count']]\n",
    "table_lengths.to_csv(f'indel_lengths_by_species.{filter_version}.cds.tsv', sep='\\t', index=False)\n",
    "table_lengths"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9a286fa",
   "metadata": {},
   "source": [
    "# Import CDS indel counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "95c14757",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Indel_type</th>\n",
       "      <th>Indel_length</th>\n",
       "      <th>mult3</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mouse</td>\n",
       "      <td>deletion</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mouse</td>\n",
       "      <td>deletion</td>\n",
       "      <td>2</td>\n",
       "      <td>.</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mouse</td>\n",
       "      <td>deletion</td>\n",
       "      <td>3</td>\n",
       "      <td>mult3</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mouse</td>\n",
       "      <td>deletion</td>\n",
       "      <td>4</td>\n",
       "      <td>.</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mouse</td>\n",
       "      <td>deletion</td>\n",
       "      <td>5</td>\n",
       "      <td>.</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>human</td>\n",
       "      <td>insertion</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>human</td>\n",
       "      <td>insertion</td>\n",
       "      <td>3</td>\n",
       "      <td>mult3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>human</td>\n",
       "      <td>insertion</td>\n",
       "      <td>6</td>\n",
       "      <td>mult3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>human</td>\n",
       "      <td>insertion</td>\n",
       "      <td>7</td>\n",
       "      <td>.</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>human</td>\n",
       "      <td>insertion</td>\n",
       "      <td>26</td>\n",
       "      <td>.</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>99 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Species Indel_type  Indel_length  mult3  Count\n",
       "0    mouse   deletion             1      .    133\n",
       "1    mouse   deletion             2      .     18\n",
       "2    mouse   deletion             3  mult3     15\n",
       "3    mouse   deletion             4      .      1\n",
       "4    mouse   deletion             5      .      5\n",
       "..     ...        ...           ...    ...    ...\n",
       "94   human  insertion             1      .     23\n",
       "95   human  insertion             3  mult3      1\n",
       "96   human  insertion             6  mult3      1\n",
       "97   human  insertion             7      .      1\n",
       "98   human  insertion            26      .      1\n",
       "\n",
       "[99 rows x 5 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "tab = pd.read_table(\"indel_lengths_by_species.including.cds.tsv\")\n",
    "tab"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cd161d8",
   "metadata": {},
   "source": [
    "# Proportion z-test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "68d976e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Indel_type</th>\n",
       "      <th>Mult3_count</th>\n",
       "      <th>Non_mult3_count</th>\n",
       "      <th>Mult3_proportion</th>\n",
       "      <th>z_statistic</th>\n",
       "      <th>p_value</th>\n",
       "      <th>Significance</th>\n",
       "      <th>Direction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Mouse</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>20</td>\n",
       "      <td>165</td>\n",
       "      <td>0.108108</td>\n",
       "      <td>-6.498441</td>\n",
       "      <td>8.115686e-11</td>\n",
       "      <td>***</td>\n",
       "      <td>Depleted</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Mouse</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>2</td>\n",
       "      <td>58</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>-4.929503</td>\n",
       "      <td>8.243906e-07</td>\n",
       "      <td>***</td>\n",
       "      <td>Depleted</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Macaque</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>135</td>\n",
       "      <td>819</td>\n",
       "      <td>0.141509</td>\n",
       "      <td>-12.568492</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>***</td>\n",
       "      <td>Depleted</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Macaque</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>16</td>\n",
       "      <td>349</td>\n",
       "      <td>0.043836</td>\n",
       "      <td>-11.732696</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>***</td>\n",
       "      <td>Depleted</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Human</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>5</td>\n",
       "      <td>34</td>\n",
       "      <td>0.128205</td>\n",
       "      <td>-2.717465</td>\n",
       "      <td>6.578414e-03</td>\n",
       "      <td>**</td>\n",
       "      <td>Depleted</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Human</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>2</td>\n",
       "      <td>25</td>\n",
       "      <td>0.074074</td>\n",
       "      <td>-2.857738</td>\n",
       "      <td>4.266725e-03</td>\n",
       "      <td>**</td>\n",
       "      <td>Depleted</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Species Indel_type  Mult3_count  Non_mult3_count  Mult3_proportion  \\\n",
       "0    Mouse   Deletion           20              165          0.108108   \n",
       "1    Mouse  Insertion            2               58          0.033333   \n",
       "2  Macaque   Deletion          135              819          0.141509   \n",
       "3  Macaque  Insertion           16              349          0.043836   \n",
       "4    Human   Deletion            5               34          0.128205   \n",
       "5    Human  Insertion            2               25          0.074074   \n",
       "\n",
       "   z_statistic       p_value Significance Direction  \n",
       "0    -6.498441  8.115686e-11          ***  Depleted  \n",
       "1    -4.929503  8.243906e-07          ***  Depleted  \n",
       "2   -12.568492  0.000000e+00          ***  Depleted  \n",
       "3   -11.732696  0.000000e+00          ***  Depleted  \n",
       "4    -2.717465  6.578414e-03           **  Depleted  \n",
       "5    -2.857738  4.266725e-03           **  Depleted  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from scipy.stats import norm\n",
    "\n",
    "# Recalculate results using one-sample proportion z-test\n",
    "results = []\n",
    "\n",
    "for species in tab['Species'].unique():\n",
    "\tfor indel_type in tab['Indel_type'].unique():\n",
    "\t\tsubset = tab[(tab['Species'] == species) & (tab['Indel_type'] == indel_type)]\n",
    "\t\t\n",
    "\t\t# Create contingency table: mult3 vs non-mult3\n",
    "\t\tmult3_count = subset[subset['mult3'] == 'mult3']['Count'].sum()\n",
    "\t\tnon_mult3_count = subset[subset['mult3'] == '.']['Count'].sum()\n",
    "\t\t\n",
    "\t\tif mult3_count + non_mult3_count > 0:\n",
    "\t\t\ttotal = mult3_count + non_mult3_count\n",
    "\t\t\tp_observed = mult3_count / total\n",
    "\t\t\tp_null = 1/3  # Expected proportion under null hypothesis\n",
    "\t\t\t\n",
    "\t\t\t# One-sample proportion z-test\n",
    "\t\t\tse = (p_null * (1 - p_null) / total) ** 0.5\n",
    "\t\t\tz_stat = (p_observed - p_null) / se\n",
    "\t\t\tp_value = 2 * (1 - norm.cdf(abs(z_stat)))  # Two-tailed test\n",
    "\t\t\t\n",
    "\t\t\tresults.append({\n",
    "\t\t\t\t'Species': species,\n",
    "\t\t\t\t'Indel_type': indel_type,\n",
    "\t\t\t\t'Mult3_count': mult3_count,\n",
    "\t\t\t\t'Non_mult3_count': non_mult3_count,\n",
    "\t\t\t\t'Mult3_proportion': p_observed,\n",
    "\t\t\t\t'z_statistic': z_stat,\n",
    "\t\t\t\t'p_value': p_value\n",
    "\t\t\t})\n",
    "\n",
    "results_df = pd.DataFrame(results)\n",
    "# Add significance column based on p-value thresholds\n",
    "results_df['Significance'] = results_df['p_value'].apply(\n",
    "\tlambda p: '***' if p < 0.001 else '**' if p < 0.01 else '*' if p < 0.05 else 'ns'\n",
    ")\n",
    "# Add enrichment direction column\n",
    "results_df['Direction'] = results_df['Mult3_proportion'].apply(\n",
    "\tlambda x: 'Enriched' if x > 1/3 else 'Depleted'\n",
    ")\n",
    "results_df['Species'] = results_df['Species'].str.capitalize()\n",
    "results_df['Indel_type'] = results_df['Indel_type'].str.capitalize()\n",
    "results_df.to_csv(\"enrichment_of_multiples_of_3_CDS_recalculated.tsv\", sep=\"\\t\", index=False)\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6880458",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
