{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3c6ca4da",
   "metadata": {},
   "source": [
    "## Parse indels and export in BED format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 778,
   "id": "d90b6b0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import sys\n",
    "\n",
    "try:\n",
    "\tspecies = sys.argv[1]  # \"human\", \"mouse\", or \"macaque\"\n",
    "\tfile_name = f\"indels.denovo_chrM_{species}.tab\"\n",
    "\tdf = pd.read_table( file_name )\n",
    "except:\n",
    "\tspecies = 'mouse'\n",
    "\tfile_name = f\"indels.denovo_chrM_{species}.tab\"\n",
    "\tdf = pd.read_table( file_name )\n",
    "\n",
    "# Import indel data.\n",
    "df = pd.read_table( file_name )\n",
    "\n",
    "# Keep only relevant columns.\n",
    "df = df[['CHROM','POS','REF','ALT','Mutation_ID']]\n",
    "df['Species'] = species\n",
    "\n",
    "# Add indel size and type.\n",
    "df['Indel_size'] = df['ALT'].str.len() - df['REF'].str.len()\n",
    "df['Indel_type'] = df['Indel_size'].apply( lambda x: 'ins' if x > 0 else 'del' )\n",
    "\n",
    "# Add BED format columns.\n",
    "df['Chrom'] = df['CHROM']\n",
    "df['Start'] = df['POS'] - 1\n",
    "df['End'] = df['POS']\n",
    "\n",
    "# Create new mutation ID.\n",
    "df['New_Mutation_ID'] = df['Species'] + '_' + df['Mutation_ID'] + '_' + df['Indel_type']\n",
    "\n",
    "# Reorder columns.\n",
    "df = df.iloc[:,-4:]\n",
    "\n",
    "# Save to BED file.\n",
    "df.to_csv( f\"{species}.indels.bed\", sep='\\t', index=False, header=False )\n",
    "\n",
    "# Split into insertions and deletions.\n",
    "df_ins = df[ df['New_Mutation_ID'].str.contains('ins') ]\n",
    "df_del = df[ df['New_Mutation_ID'].str.contains('del') ]\n",
    "# Save to separate BED files.\n",
    "df_ins.to_csv( f\"{species}.indels.ins.bed\", sep='\\t', index=False, header=False )\n",
    "df_del.to_csv( f\"{species}.indels.del.bed\", sep='\\t', index=False, header=False )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6840943",
   "metadata": {},
   "source": [
    "## Get flanking sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "290230df",
   "metadata": {},
   "source": [
    "### Run bedtools slop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 779,
   "id": "ce070430",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Created genome.txt for mouse chrM\n",
      "Created mouse.indels.del.slop200.bed with 200 bp flanks\n",
      "Created mouse.indels.ins.slop200.bed with 200 bp flanks\n",
      "Created genome.txt for macaque chrM\n",
      "Created macaque.indels.del.slop200.bed with 200 bp flanks\n",
      "Created macaque.indels.ins.slop200.bed with 200 bp flanks\n",
      "Created genome.txt for human chrM\n",
      "Created human.indels.del.slop200.bed with 200 bp flanks\n",
      "Created human.indels.ins.slop200.bed with 200 bp flanks\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "slop_size = 200  # Number of base pairs to add as flanks\n",
    "species_list = [\"mouse\", \"macaque\", \"human\"]\n",
    "\n",
    "for species in species_list:\n",
    "\tif species == \"mouse\":\n",
    "\t\tgenome_size = 16300\n",
    "\telif species == \"macaque\":\n",
    "\t\tgenome_size = 16564\n",
    "\telif species == \"human\":\n",
    "\t\tgenome_size = 16569\n",
    "\telse:\n",
    "\t\tprint(f\"Unknown species: {species}\")\n",
    "\t\tcontinue\n",
    "\n",
    "\twith open(\"genome.txt\", \"w\") as f:\n",
    "\t\tf.write(f\"chrM\\t{genome_size}\\n\")\n",
    "\tprint(f\"Created genome.txt for {species} chrM\")\n",
    "\n",
    "\t# Deletions\n",
    "\tsubprocess.run([\n",
    "\t\t\"bedtools\", \"slop\",\n",
    "\t\t\"-i\", f\"{species}.indels.del.bed\",\n",
    "\t\t\"-b\", f\"{slop_size}\",\n",
    "\t\t\"-g\", \"genome.txt\"\n",
    "\t], stdout=open(f\"{species}.indels.del.slop{slop_size}.bed\", \"w\"))\n",
    "\tprint(f\"Created {species}.indels.del.slop{slop_size}.bed with {slop_size} bp flanks\")\n",
    "\n",
    "\t# Insertions\n",
    "\tsubprocess.run([\n",
    "\t\t\"bedtools\", \"slop\",\n",
    "\t\t\"-i\", f\"{species}.indels.ins.bed\",\n",
    "\t\t\"-b\", f\"{slop_size}\",\n",
    "\t\t\"-g\", \"genome.txt\"\n",
    "\t], stdout=open(f\"{species}.indels.ins.slop{slop_size}.bed\", \"w\"))\n",
    "\tprint(f\"Created {species}.indels.ins.slop{slop_size}.bed with {slop_size} bp flanks\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93d3c6c7",
   "metadata": {},
   "source": [
    "### Run bedtools getfasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 780,
   "id": "dcc85d49",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Created mouse.indels.ins.slop200.fa with sequences\n",
      "Created mouse.indels.del.slop200.fa with sequences\n",
      "Created macaque.indels.ins.slop200.fa with sequences\n",
      "Created macaque.indels.del.slop200.fa with sequences\n",
      "Created human.indels.ins.slop200.fa with sequences\n",
      "Created human.indels.del.slop200.fa with sequences\n"
     ]
    }
   ],
   "source": [
    "fasta_dir = \"../../data/refs/\"\n",
    "fasta_files = {\n",
    "\t\"mouse\": f\"{fasta_dir}/mouse/NC_005089_9821insA.chrM.fa\",\n",
    "\t\"macaque\": f\"{fasta_dir}/rhesus_macaque/rheMac10.chrM.fasta\",\n",
    "\t\"human\": f\"{fasta_dir}/human/NC_012920.1.fa\"\n",
    "}\n",
    "\n",
    "for sp in species_list:\n",
    "\tfasta_file = fasta_files[sp]\n",
    "\t# Insertions\n",
    "\tsubprocess.run([\n",
    "\t\t\"bedtools\", \"getfasta\",\n",
    "\t\t\"-bed\", f\"{sp}.indels.ins.slop{slop_size}.bed\",\n",
    "\t\t\"-fi\", fasta_file,\n",
    "\t\t\"-fo\", f\"{sp}.indels.ins.slop{slop_size}.fa\",\n",
    "\t\t\"-name+\"\n",
    "\t])\n",
    "\tprint(f\"Created {sp}.indels.ins.slop{slop_size}.fa with sequences\")\n",
    "\t# Deletions\n",
    "\tsubprocess.run([\n",
    "\t\t\"bedtools\", \"getfasta\",\n",
    "\t\t\"-bed\", f\"{sp}.indels.del.slop{slop_size}.bed\",\n",
    "\t\t\"-fi\", fasta_file,\n",
    "\t\t\"-fo\", f\"{sp}.indels.del.slop{slop_size}.fa\",\n",
    "\t\t\"-name+\"\n",
    "\t])\n",
    "\tprint(f\"Created {sp}.indels.del.slop{slop_size}.fa with sequences\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf9cb970",
   "metadata": {},
   "source": [
    "# Insertions:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4d43992",
   "metadata": {},
   "source": [
    "## Parse FastA file sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 781,
   "id": "800a9a5e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fasta_ID</th>\n",
       "      <th>Left_Flank</th>\n",
       "      <th>Right_Flank</th>\n",
       "      <th>Mutation_ID</th>\n",
       "      <th>Left_Flank_flipped</th>\n",
       "      <th>Species</th>\n",
       "      <th>Position</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Type</th>\n",
       "      <th>Region</th>\n",
       "      <th>Inserted_bases</th>\n",
       "      <th>Inserted_bases_len</th>\n",
       "      <th>Left_Flank_flipped_shortened</th>\n",
       "      <th>Right_Flank_shortened</th>\n",
       "      <th>Right_Flank_shortened_flipped</th>\n",
       "      <th>Left_Flank_flipped_shortened_unflipped</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mouse_13053_T_TC_ins::chrM:12852-13253</td>\n",
       "      <td>TCATGCCTAGTAATCGGAAGCCTCGCCCTCACAGGAATACCATTCC...</td>\n",
       "      <td>CCCCCCCTAATCTCCATTAACGAAAATGACCCAGACCTCATAAACC...</td>\n",
       "      <td>13053_T_TC</td>\n",
       "      <td>TTTTGCGCCAAAACAATAACAATGCTTCATTTACTAAGCATACGAC...</td>\n",
       "      <td>mouse</td>\n",
       "      <td>13053</td>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "      <td>ins</td>\n",
       "      <td>chrM:12852-13253</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mouse_14640_A_AG_ins::chrM:14439-14840</td>\n",
       "      <td>GGACGAGGCTTATATTATGGATCATATACATTTATAGAAACCTGAA...</td>\n",
       "      <td>GGGGGCTTCTCAGTAGACAAAGCCACCTTGACCCGATTCTTCGCTT...</td>\n",
       "      <td>14640_A_AG</td>\n",
       "      <td>AGTTTAAGTAAGCTGATCCCAACAAGGTTATATACCCTACCGACTA...</td>\n",
       "      <td>mouse</td>\n",
       "      <td>14640</td>\n",
       "      <td>A</td>\n",
       "      <td>AG</td>\n",
       "      <td>ins</td>\n",
       "      <td>chrM:14439-14840</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mouse_5171_G_GA_ins::chrM:4970-5371</td>\n",
       "      <td>CCGCGAGCCTTCAAAGCCCTAAGAAAACACACAAGTTTAACTTCTG...</td>\n",
       "      <td>AAAAAAAAAAATGGCGGTAGAAGTCTTAGTAGAGATTTCTCTACAC...</td>\n",
       "      <td>5171_G_GA</td>\n",
       "      <td>GCCGCCATCTTCATCTAACTTCGGTCATTATCCCATAAATCGACAA...</td>\n",
       "      <td>mouse</td>\n",
       "      <td>5171</td>\n",
       "      <td>G</td>\n",
       "      <td>GA</td>\n",
       "      <td>ins</td>\n",
       "      <td>chrM:4970-5371</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mouse_9841_G_GT_ins::chrM:9640-10041</td>\n",
       "      <td>ATTTCTATTATTTGACCTAGAAATTGCTCTTCTACTTCCACTACCA...</td>\n",
       "      <td>ACTCATTAGATTATGATGATGTTCATAATTACCAATATGCCATCTA...</td>\n",
       "      <td>9841_G_GT</td>\n",
       "      <td>GCTTTAGTAATTAAAAAAAAATTTGATTAATGGTAAATGAGACAAG...</td>\n",
       "      <td>mouse</td>\n",
       "      <td>9841</td>\n",
       "      <td>G</td>\n",
       "      <td>GT</td>\n",
       "      <td>ins</td>\n",
       "      <td>chrM:9640-10041</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mouse_12636_C_CA_ins::chrM:12435-12836</td>\n",
       "      <td>TGACTACCATCAGCAATAGAAGGCCCTACACCAGTTTCAGCACTAC...</td>\n",
       "      <td>AAAAAAATCATTGCCTTCTCTACATCAAGCCAACTAGGCCTGATAA...</td>\n",
       "      <td>12636_C_CA</td>\n",
       "      <td>CTACAGCAAAACCCACTCTCGTGTTTATCGACATTTATTACACCAA...</td>\n",
       "      <td>mouse</td>\n",
       "      <td>12636</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>ins</td>\n",
       "      <td>chrM:12435-12836</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>445</th>\n",
       "      <td>human_16182_A_AC_ins::chrM:15981-16382</td>\n",
       "      <td>ATTAGCACCCAAAGCTAAGATTCTAATTTAAACTATTCTCTGTTCT...</td>\n",
       "      <td>ACCCCCTCCCCATGCTTACAAGCAAGTACAGCAATCAACCCTCAAC...</td>\n",
       "      <td>16182_A_AC</td>\n",
       "      <td>AAACTACACCTAACCCAAAAATACATGATGTCCACCAGTTCATAAA...</td>\n",
       "      <td>human</td>\n",
       "      <td>16182</td>\n",
       "      <td>A</td>\n",
       "      <td>AC</td>\n",
       "      <td>ins</td>\n",
       "      <td>chrM:15981-16382</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>446</th>\n",
       "      <td>human_16188_T_TC_ins::chrM:15987-16388</td>\n",
       "      <td>ACCCAAAGCTAAGATTCTAATTTAAACTATTCTCTGTTCTTTCATG...</td>\n",
       "      <td>TCCCCATGCTTACAAGCAAGTACAGCAATCAACCCTCAACTATCAC...</td>\n",
       "      <td>16188_T_TC</td>\n",
       "      <td>CCCCCAAAACTACACCTAACCCAAAAATACATGATGTCCACCAGTT...</td>\n",
       "      <td>human</td>\n",
       "      <td>16188</td>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "      <td>ins</td>\n",
       "      <td>chrM:15987-16388</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447</th>\n",
       "      <td>human_16188_T_TCC_ins::chrM:15987-16388</td>\n",
       "      <td>ACCCAAAGCTAAGATTCTAATTTAAACTATTCTCTGTTCTTTCATG...</td>\n",
       "      <td>TCCCCATGCTTACAAGCAAGTACAGCAATCAACCCTCAACTATCAC...</td>\n",
       "      <td>16188_T_TCC</td>\n",
       "      <td>CCCCCAAAACTACACCTAACCCAAAAATACATGATGTCCACCAGTT...</td>\n",
       "      <td>human</td>\n",
       "      <td>16188</td>\n",
       "      <td>T</td>\n",
       "      <td>TCC</td>\n",
       "      <td>ins</td>\n",
       "      <td>chrM:15987-16388</td>\n",
       "      <td>CC</td>\n",
       "      <td>2</td>\n",
       "      <td>CC</td>\n",
       "      <td>TC</td>\n",
       "      <td>CT</td>\n",
       "      <td>CC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>448</th>\n",
       "      <td>human_16356_T_TCTCGTCCCCATGGATGACCCCC_ins::chr...</td>\n",
       "      <td>GTACATAAAAACCCAATCCACATCAAAACCCCCTCCCCATGCTTAC...</td>\n",
       "      <td>TCTCGTCCCCATGGATGACCCCCCTCAGATAGGGGTCCCTTGACCA...</td>\n",
       "      <td>16356_T_TCTCGTCCCCATGGATGACCCCC</td>\n",
       "      <td>TCCCTAAACTGACATTACACGATACATGCCATTTACCGAAATACAT...</td>\n",
       "      <td>human</td>\n",
       "      <td>16356</td>\n",
       "      <td>T</td>\n",
       "      <td>TCTCGTCCCCATGGATGACCCCC</td>\n",
       "      <td>ins</td>\n",
       "      <td>chrM:16155-16556</td>\n",
       "      <td>CTCGTCCCCATGGATGACCCCC</td>\n",
       "      <td>22</td>\n",
       "      <td>TCCCTAAACTGACATTACACGA</td>\n",
       "      <td>TCTCGTCCCCATGGATGACCCC</td>\n",
       "      <td>CCCCAGTAGGTACCCCTGCTCT</td>\n",
       "      <td>AGCACATTACAGTCAAATCCCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>449</th>\n",
       "      <td>human_16373_A_AC_ins::chrM:16172-16569</td>\n",
       "      <td>CCACATCAAAACCCCCTCCCCATGCTTACAAGCAAGTACAGCAATC...</td>\n",
       "      <td>TGACCCCCCTCAGATAGGGGTCCCTTGACCACCATCCTCCGTGAAA...</td>\n",
       "      <td>16373_A_AC</td>\n",
       "      <td>AGGTACCCCTGCTCTTCCCTAAACTGACATTACACGATACATGCCA...</td>\n",
       "      <td>human</td>\n",
       "      <td>16373</td>\n",
       "      <td>A</td>\n",
       "      <td>AC</td>\n",
       "      <td>ins</td>\n",
       "      <td>chrM:16172-16569</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>450 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              Fasta_ID  \\\n",
       "0               mouse_13053_T_TC_ins::chrM:12852-13253   \n",
       "1               mouse_14640_A_AG_ins::chrM:14439-14840   \n",
       "2                  mouse_5171_G_GA_ins::chrM:4970-5371   \n",
       "3                 mouse_9841_G_GT_ins::chrM:9640-10041   \n",
       "4               mouse_12636_C_CA_ins::chrM:12435-12836   \n",
       "..                                                 ...   \n",
       "445             human_16182_A_AC_ins::chrM:15981-16382   \n",
       "446             human_16188_T_TC_ins::chrM:15987-16388   \n",
       "447            human_16188_T_TCC_ins::chrM:15987-16388   \n",
       "448  human_16356_T_TCTCGTCCCCATGGATGACCCCC_ins::chr...   \n",
       "449             human_16373_A_AC_ins::chrM:16172-16569   \n",
       "\n",
       "                                            Left_Flank  \\\n",
       "0    TCATGCCTAGTAATCGGAAGCCTCGCCCTCACAGGAATACCATTCC...   \n",
       "1    GGACGAGGCTTATATTATGGATCATATACATTTATAGAAACCTGAA...   \n",
       "2    CCGCGAGCCTTCAAAGCCCTAAGAAAACACACAAGTTTAACTTCTG...   \n",
       "3    ATTTCTATTATTTGACCTAGAAATTGCTCTTCTACTTCCACTACCA...   \n",
       "4    TGACTACCATCAGCAATAGAAGGCCCTACACCAGTTTCAGCACTAC...   \n",
       "..                                                 ...   \n",
       "445  ATTAGCACCCAAAGCTAAGATTCTAATTTAAACTATTCTCTGTTCT...   \n",
       "446  ACCCAAAGCTAAGATTCTAATTTAAACTATTCTCTGTTCTTTCATG...   \n",
       "447  ACCCAAAGCTAAGATTCTAATTTAAACTATTCTCTGTTCTTTCATG...   \n",
       "448  GTACATAAAAACCCAATCCACATCAAAACCCCCTCCCCATGCTTAC...   \n",
       "449  CCACATCAAAACCCCCTCCCCATGCTTACAAGCAAGTACAGCAATC...   \n",
       "\n",
       "                                           Right_Flank  \\\n",
       "0    CCCCCCCTAATCTCCATTAACGAAAATGACCCAGACCTCATAAACC...   \n",
       "1    GGGGGCTTCTCAGTAGACAAAGCCACCTTGACCCGATTCTTCGCTT...   \n",
       "2    AAAAAAAAAAATGGCGGTAGAAGTCTTAGTAGAGATTTCTCTACAC...   \n",
       "3    ACTCATTAGATTATGATGATGTTCATAATTACCAATATGCCATCTA...   \n",
       "4    AAAAAAATCATTGCCTTCTCTACATCAAGCCAACTAGGCCTGATAA...   \n",
       "..                                                 ...   \n",
       "445  ACCCCCTCCCCATGCTTACAAGCAAGTACAGCAATCAACCCTCAAC...   \n",
       "446  TCCCCATGCTTACAAGCAAGTACAGCAATCAACCCTCAACTATCAC...   \n",
       "447  TCCCCATGCTTACAAGCAAGTACAGCAATCAACCCTCAACTATCAC...   \n",
       "448  TCTCGTCCCCATGGATGACCCCCCTCAGATAGGGGTCCCTTGACCA...   \n",
       "449  TGACCCCCCTCAGATAGGGGTCCCTTGACCACCATCCTCCGTGAAA...   \n",
       "\n",
       "                         Mutation_ID  \\\n",
       "0                         13053_T_TC   \n",
       "1                         14640_A_AG   \n",
       "2                          5171_G_GA   \n",
       "3                          9841_G_GT   \n",
       "4                         12636_C_CA   \n",
       "..                               ...   \n",
       "445                       16182_A_AC   \n",
       "446                       16188_T_TC   \n",
       "447                      16188_T_TCC   \n",
       "448  16356_T_TCTCGTCCCCATGGATGACCCCC   \n",
       "449                       16373_A_AC   \n",
       "\n",
       "                                    Left_Flank_flipped Species Position Ref  \\\n",
       "0    TTTTGCGCCAAAACAATAACAATGCTTCATTTACTAAGCATACGAC...   mouse    13053   T   \n",
       "1    AGTTTAAGTAAGCTGATCCCAACAAGGTTATATACCCTACCGACTA...   mouse    14640   A   \n",
       "2    GCCGCCATCTTCATCTAACTTCGGTCATTATCCCATAAATCGACAA...   mouse     5171   G   \n",
       "3    GCTTTAGTAATTAAAAAAAAATTTGATTAATGGTAAATGAGACAAG...   mouse     9841   G   \n",
       "4    CTACAGCAAAACCCACTCTCGTGTTTATCGACATTTATTACACCAA...   mouse    12636   C   \n",
       "..                                                 ...     ...      ...  ..   \n",
       "445  AAACTACACCTAACCCAAAAATACATGATGTCCACCAGTTCATAAA...   human    16182   A   \n",
       "446  CCCCCAAAACTACACCTAACCCAAAAATACATGATGTCCACCAGTT...   human    16188   T   \n",
       "447  CCCCCAAAACTACACCTAACCCAAAAATACATGATGTCCACCAGTT...   human    16188   T   \n",
       "448  TCCCTAAACTGACATTACACGATACATGCCATTTACCGAAATACAT...   human    16356   T   \n",
       "449  AGGTACCCCTGCTCTTCCCTAAACTGACATTACACGATACATGCCA...   human    16373   A   \n",
       "\n",
       "                         Alt Type            Region          Inserted_bases  \\\n",
       "0                         TC  ins  chrM:12852-13253                       C   \n",
       "1                         AG  ins  chrM:14439-14840                       G   \n",
       "2                         GA  ins    chrM:4970-5371                       A   \n",
       "3                         GT  ins   chrM:9640-10041                       T   \n",
       "4                         CA  ins  chrM:12435-12836                       A   \n",
       "..                       ...  ...               ...                     ...   \n",
       "445                       AC  ins  chrM:15981-16382                       C   \n",
       "446                       TC  ins  chrM:15987-16388                       C   \n",
       "447                      TCC  ins  chrM:15987-16388                      CC   \n",
       "448  TCTCGTCCCCATGGATGACCCCC  ins  chrM:16155-16556  CTCGTCCCCATGGATGACCCCC   \n",
       "449                       AC  ins  chrM:16172-16569                       C   \n",
       "\n",
       "     Inserted_bases_len Left_Flank_flipped_shortened   Right_Flank_shortened  \\\n",
       "0                     1                            T                       C   \n",
       "1                     1                            A                       G   \n",
       "2                     1                            G                       A   \n",
       "3                     1                            G                       A   \n",
       "4                     1                            C                       A   \n",
       "..                  ...                          ...                     ...   \n",
       "445                   1                            A                       A   \n",
       "446                   1                            C                       T   \n",
       "447                   2                           CC                      TC   \n",
       "448                  22       TCCCTAAACTGACATTACACGA  TCTCGTCCCCATGGATGACCCC   \n",
       "449                   1                            A                       T   \n",
       "\n",
       "    Right_Flank_shortened_flipped Left_Flank_flipped_shortened_unflipped  \n",
       "0                               C                                      T  \n",
       "1                               G                                      A  \n",
       "2                               A                                      G  \n",
       "3                               A                                      G  \n",
       "4                               A                                      C  \n",
       "..                            ...                                    ...  \n",
       "445                             A                                      A  \n",
       "446                             T                                      C  \n",
       "447                            CT                                     CC  \n",
       "448        CCCCAGTAGGTACCCCTGCTCT                 AGCACATTACAGTCAAATCCCT  \n",
       "449                             T                                      A  \n",
       "\n",
       "[450 rows x 17 columns]"
      ]
     },
     "execution_count": 781,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_rows', 20)\n",
    "\n",
    "all_tables = []\n",
    "\n",
    "for species in species_list:\n",
    "\tfasta_file = f\"{species}.indels.ins.slop{slop_size}.fa\"\n",
    "\tfasta_dict = {record.id: str(record.seq) for record in SeqIO.parse(fasta_file, \"fasta\")}\n",
    "\n",
    "\tsplit_fasta_dict = {}\n",
    "\tfor key, seq in fasta_dict.items():\n",
    "\t\tmid = len(seq) // 2\n",
    "\t\tsplit_fasta_dict[key] = (seq[:mid+1], seq[mid+1:])\n",
    "\n",
    "\ttable = pd.DataFrame([\n",
    "\t\t{'Fasta_ID': k, 'Left_Flank': v[0], 'Right_Flank': v[1]}\n",
    "\t\tfor k, v in split_fasta_dict.items()\n",
    "\t])\n",
    "\t\n",
    "\t# Extract Mutation_ID from Fasta_ID\n",
    "\ttable['Mutation_ID'] = table['Fasta_ID'].str.split(':').str[0].str.split('_').str[1:4].str.join('_')\n",
    "\n",
    "\ttable['Left_Flank'] = table['Left_Flank'].str.upper()\n",
    "\ttable['Right_Flank'] = table['Right_Flank'].str.upper()\n",
    "\ttable['Left_Flank_flipped'] = table['Left_Flank'].apply(lambda x: x[::-1])\n",
    "\ttable[['Species', 'Position', 'Ref', 'Alt', 'Type', 'Region']] = table['Fasta_ID'].str.extract(r'([^_]+)_([^_]+)_([^_]+)_([^_]+)_([^:]+)::(.+)')\n",
    "\ttable['Inserted_bases'] = table['Alt'].str[1:]\n",
    "\ttable['Inserted_bases_len'] = table['Inserted_bases'].str.len()\n",
    "\n",
    "\t# Consider only the flank that is the same size as the insertion.\n",
    "\ttable['Left_Flank_flipped_shortened'] = table.apply(lambda row: \n",
    "\t\t\t\t\t\t\t\t\trow['Left_Flank_flipped'][:row['Inserted_bases_len']], axis=1)\n",
    "\ttable['Right_Flank_shortened'] = table.apply(lambda row: \n",
    "\t\t\t\t\t\t\t\t\trow['Right_Flank'][:row['Inserted_bases_len']], axis=1)\n",
    "\ttable['Right_Flank_shortened_flipped'] = table['Right_Flank_shortened'].apply(lambda x: x[::-1])\n",
    "\ttable['Left_Flank_flipped_shortened_unflipped'] = table['Left_Flank_flipped_shortened'].apply(lambda x: x[::-1])\n",
    "\n",
    "\tall_tables.append(table)\n",
    "\n",
    "# Concatenate all tables into one DataFrame\n",
    "table = pd.concat(all_tables, ignore_index=True)\n",
    "table\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5292a8d5",
   "metadata": {},
   "source": [
    "### Homopolymer?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 782,
   "id": "86227709",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inserted_bases</th>\n",
       "      <th>Right_Flank</th>\n",
       "      <th>Right_Flank_homopolymer</th>\n",
       "      <th>Left_Flank_flipped</th>\n",
       "      <th>Left_Flank_flipped_homopolymer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>CCCCCCCTAATCTCCATTAACGAAAATGACCCAGACCTCATAAACC...</td>\n",
       "      <td>True</td>\n",
       "      <td>TTTTGCGCCAAAACAATAACAATGCTTCATTTACTAAGCATACGAC...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G</td>\n",
       "      <td>GGGGGCTTCTCAGTAGACAAAGCCACCTTGACCCGATTCTTCGCTT...</td>\n",
       "      <td>True</td>\n",
       "      <td>AGTTTAAGTAAGCTGATCCCAACAAGGTTATATACCCTACCGACTA...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A</td>\n",
       "      <td>AAAAAAAAAAATGGCGGTAGAAGTCTTAGTAGAGATTTCTCTACAC...</td>\n",
       "      <td>True</td>\n",
       "      <td>GCCGCCATCTTCATCTAACTTCGGTCATTATCCCATAAATCGACAA...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T</td>\n",
       "      <td>ACTCATTAGATTATGATGATGTTCATAATTACCAATATGCCATCTA...</td>\n",
       "      <td>False</td>\n",
       "      <td>GCTTTAGTAATTAAAAAAAAATTTGATTAATGGTAAATGAGACAAG...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A</td>\n",
       "      <td>AAAAAAATCATTGCCTTCTCTACATCAAGCCAACTAGGCCTGATAA...</td>\n",
       "      <td>True</td>\n",
       "      <td>CTACAGCAAAACCCACTCTCGTGTTTATCGACATTTATTACACCAA...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>445</th>\n",
       "      <td>C</td>\n",
       "      <td>ACCCCCTCCCCATGCTTACAAGCAAGTACAGCAATCAACCCTCAAC...</td>\n",
       "      <td>False</td>\n",
       "      <td>AAACTACACCTAACCCAAAAATACATGATGTCCACCAGTTCATAAA...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>446</th>\n",
       "      <td>C</td>\n",
       "      <td>TCCCCATGCTTACAAGCAAGTACAGCAATCAACCCTCAACTATCAC...</td>\n",
       "      <td>False</td>\n",
       "      <td>CCCCCAAAACTACACCTAACCCAAAAATACATGATGTCCACCAGTT...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447</th>\n",
       "      <td>CC</td>\n",
       "      <td>TCCCCATGCTTACAAGCAAGTACAGCAATCAACCCTCAACTATCAC...</td>\n",
       "      <td>False</td>\n",
       "      <td>CCCCCAAAACTACACCTAACCCAAAAATACATGATGTCCACCAGTT...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>448</th>\n",
       "      <td>CTCGTCCCCATGGATGACCCCC</td>\n",
       "      <td>TCTCGTCCCCATGGATGACCCCCCTCAGATAGGGGTCCCTTGACCA...</td>\n",
       "      <td>False</td>\n",
       "      <td>TCCCTAAACTGACATTACACGATACATGCCATTTACCGAAATACAT...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>449</th>\n",
       "      <td>C</td>\n",
       "      <td>TGACCCCCCTCAGATAGGGGTCCCTTGACCACCATCCTCCGTGAAA...</td>\n",
       "      <td>False</td>\n",
       "      <td>AGGTACCCCTGCTCTTCCCTAAACTGACATTACACGATACATGCCA...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>450 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Inserted_bases  \\\n",
       "0                         C   \n",
       "1                         G   \n",
       "2                         A   \n",
       "3                         T   \n",
       "4                         A   \n",
       "..                      ...   \n",
       "445                       C   \n",
       "446                       C   \n",
       "447                      CC   \n",
       "448  CTCGTCCCCATGGATGACCCCC   \n",
       "449                       C   \n",
       "\n",
       "                                           Right_Flank  \\\n",
       "0    CCCCCCCTAATCTCCATTAACGAAAATGACCCAGACCTCATAAACC...   \n",
       "1    GGGGGCTTCTCAGTAGACAAAGCCACCTTGACCCGATTCTTCGCTT...   \n",
       "2    AAAAAAAAAAATGGCGGTAGAAGTCTTAGTAGAGATTTCTCTACAC...   \n",
       "3    ACTCATTAGATTATGATGATGTTCATAATTACCAATATGCCATCTA...   \n",
       "4    AAAAAAATCATTGCCTTCTCTACATCAAGCCAACTAGGCCTGATAA...   \n",
       "..                                                 ...   \n",
       "445  ACCCCCTCCCCATGCTTACAAGCAAGTACAGCAATCAACCCTCAAC...   \n",
       "446  TCCCCATGCTTACAAGCAAGTACAGCAATCAACCCTCAACTATCAC...   \n",
       "447  TCCCCATGCTTACAAGCAAGTACAGCAATCAACCCTCAACTATCAC...   \n",
       "448  TCTCGTCCCCATGGATGACCCCCCTCAGATAGGGGTCCCTTGACCA...   \n",
       "449  TGACCCCCCTCAGATAGGGGTCCCTTGACCACCATCCTCCGTGAAA...   \n",
       "\n",
       "     Right_Flank_homopolymer  \\\n",
       "0                       True   \n",
       "1                       True   \n",
       "2                       True   \n",
       "3                      False   \n",
       "4                       True   \n",
       "..                       ...   \n",
       "445                    False   \n",
       "446                    False   \n",
       "447                    False   \n",
       "448                    False   \n",
       "449                    False   \n",
       "\n",
       "                                    Left_Flank_flipped  \\\n",
       "0    TTTTGCGCCAAAACAATAACAATGCTTCATTTACTAAGCATACGAC...   \n",
       "1    AGTTTAAGTAAGCTGATCCCAACAAGGTTATATACCCTACCGACTA...   \n",
       "2    GCCGCCATCTTCATCTAACTTCGGTCATTATCCCATAAATCGACAA...   \n",
       "3    GCTTTAGTAATTAAAAAAAAATTTGATTAATGGTAAATGAGACAAG...   \n",
       "4    CTACAGCAAAACCCACTCTCGTGTTTATCGACATTTATTACACCAA...   \n",
       "..                                                 ...   \n",
       "445  AAACTACACCTAACCCAAAAATACATGATGTCCACCAGTTCATAAA...   \n",
       "446  CCCCCAAAACTACACCTAACCCAAAAATACATGATGTCCACCAGTT...   \n",
       "447  CCCCCAAAACTACACCTAACCCAAAAATACATGATGTCCACCAGTT...   \n",
       "448  TCCCTAAACTGACATTACACGATACATGCCATTTACCGAAATACAT...   \n",
       "449  AGGTACCCCTGCTCTTCCCTAAACTGACATTACACGATACATGCCA...   \n",
       "\n",
       "     Left_Flank_flipped_homopolymer  \n",
       "0                              True  \n",
       "1                             False  \n",
       "2                             False  \n",
       "3                             False  \n",
       "4                             False  \n",
       "..                              ...  \n",
       "445                            True  \n",
       "446                            True  \n",
       "447                            True  \n",
       "448                           False  \n",
       "449                           False  \n",
       "\n",
       "[450 rows x 5 columns]"
      ]
     },
     "execution_count": 782,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check if the beginning of Right_Flank is a homopolymer (e.g., first 2-5 bases are the same)\n",
    "def is_homopolymer(seq, n=3):\n",
    "\t\"\"\"Return True if the first n bases of seq are all the same.\"\"\"\n",
    "\treturn len(seq) >= n and seq[:n] == seq[0] * n\n",
    "\n",
    "# Apply to the Right_Flank column for n=3 (adjust n as needed)\n",
    "table['Right_Flank_homopolymer'] = table['Right_Flank'].apply(lambda x: is_homopolymer(x, n=3))\n",
    "table[['Right_Flank', 'Right_Flank_homopolymer']]\n",
    "\n",
    "# Check if the beginning of Left_Flank_flipped is a homopolymer (e.g., first 3 bases are the same)\n",
    "table['Left_Flank_flipped'] = table['Left_Flank_flipped']\n",
    "table['Left_Flank_flipped_homopolymer'] = table['Left_Flank_flipped'].apply(lambda x: is_homopolymer(x, n=3))\n",
    "\n",
    "# Show both Right_Flank_homopolymer and Left_Flank_flipped_homopolymer in the same table\n",
    "table[['Inserted_bases','Right_Flank', 'Right_Flank_homopolymer', 'Left_Flank_flipped', 'Left_Flank_flipped_homopolymer']]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d339d80a",
   "metadata": {},
   "source": [
    "### Hamming distance between inserted bases and flanks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 783,
   "id": "d3eae612",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Left_Flank_flipped_shortened_unflipped</th>\n",
       "      <th>Inserted_bases</th>\n",
       "      <th>Right_Flank_shortened</th>\n",
       "      <th>Hamming_Left</th>\n",
       "      <th>Hamming_Right</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>445</th>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>446</th>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447</th>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>TC</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>448</th>\n",
       "      <td>AGCACATTACAGTCAAATCCCT</td>\n",
       "      <td>CTCGTCCCCATGGATGACCCCC</td>\n",
       "      <td>TCTCGTCCCCATGGATGACCCC</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>0.363636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>449</th>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>450 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Left_Flank_flipped_shortened_unflipped          Inserted_bases  \\\n",
       "0                                        T                       C   \n",
       "1                                        A                       G   \n",
       "2                                        G                       A   \n",
       "3                                        G                       T   \n",
       "4                                        C                       A   \n",
       "..                                     ...                     ...   \n",
       "445                                      A                       C   \n",
       "446                                      C                       C   \n",
       "447                                     CC                      CC   \n",
       "448                 AGCACATTACAGTCAAATCCCT  CTCGTCCCCATGGATGACCCCC   \n",
       "449                                      A                       C   \n",
       "\n",
       "      Right_Flank_shortened  Hamming_Left  Hamming_Right  \n",
       "0                         C      0.000000       1.000000  \n",
       "1                         G      0.000000       1.000000  \n",
       "2                         A      0.000000       1.000000  \n",
       "3                         A      0.000000       0.000000  \n",
       "4                         A      0.000000       1.000000  \n",
       "..                      ...           ...            ...  \n",
       "445                       A      0.000000       0.000000  \n",
       "446                       T      1.000000       0.000000  \n",
       "447                      TC      1.000000       0.500000  \n",
       "448  TCTCGTCCCCATGGATGACCCC      0.272727       0.363636  \n",
       "449                       T      0.000000       0.000000  \n",
       "\n",
       "[450 rows x 5 columns]"
      ]
     },
     "execution_count": 783,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from difflib import SequenceMatcher\n",
    "\n",
    "def hamming_similarity(s1, s2):\n",
    "\t\"\"\"Return the proportion of matching characters (Hamming similarity) between two strings of equal length.\"\"\"\n",
    "\tif len(s1) != len(s2) or len(s1) == 0:\n",
    "\t\treturn None\n",
    "\treturn sum(a == b for a, b in zip(s1, s2)) / len(s1)\n",
    "\n",
    "# Compute Hamming similarity for each row between Inserted_bases and Left_Flank_flipped_shortened, Right_Flank_shortened.\n",
    "table['Hamming_Left'] = table.apply(\n",
    "\tlambda row: hamming_similarity(row['Inserted_bases'], row['Left_Flank_flipped_shortened_unflipped']), axis=1\n",
    ")\n",
    "\n",
    "# Compute Hamming similarity for each row between Inserted_bases and Right_Flank_shortened.\n",
    "table['Hamming_Right'] = table.apply(\n",
    "\tlambda row: hamming_similarity(row['Inserted_bases'], row['Right_Flank_shortened']), axis=1\n",
    ")\n",
    "\n",
    "# Identify rows where the deleted bases are identical to either flank.\n",
    "table['Identical_to_flank'] = (table['Hamming_Left'] == 1) | (table['Hamming_Right'] == 1)\n",
    "\n",
    "# Create a copy of the table for insertions.\n",
    "table_insertions = table.copy()\n",
    "\n",
    "table[['Left_Flank_flipped_shortened_unflipped', 'Inserted_bases', 'Right_Flank_shortened', 'Hamming_Left', 'Hamming_Right']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 784,
   "id": "b96cff30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Left_Flank_flipped_shortened_unflipped</th>\n",
       "      <th>Inserted_bases</th>\n",
       "      <th>Right_Flank_shortened</th>\n",
       "      <th>Hamming_Left</th>\n",
       "      <th>Hamming_Right</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>389</th>\n",
       "      <td>CCCC</td>\n",
       "      <td>CCCT</td>\n",
       "      <td>CTCC</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>390</th>\n",
       "      <td>CCCC</td>\n",
       "      <td>CCTC</td>\n",
       "      <td>TCCC</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>392</th>\n",
       "      <td>CCC</td>\n",
       "      <td>CTC</td>\n",
       "      <td>TCC</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>AAG</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTG</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394</th>\n",
       "      <td>ATACT</td>\n",
       "      <td>ATGTT</td>\n",
       "      <td>AACCC</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>427</th>\n",
       "      <td>TAGAATT</td>\n",
       "      <td>ATTCCCC</td>\n",
       "      <td>AATTCCC</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.714286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>AAG</td>\n",
       "      <td>CCA</td>\n",
       "      <td>TTA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>TGT</td>\n",
       "      <td>AAC</td>\n",
       "      <td>ACG</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>ACA</td>\n",
       "      <td>TTC</td>\n",
       "      <td>ACC</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419</th>\n",
       "      <td>TAGACG</td>\n",
       "      <td>AGACAC</td>\n",
       "      <td>TAGACA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>49 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Left_Flank_flipped_shortened_unflipped Inserted_bases  \\\n",
       "389                                   CCCC           CCCT   \n",
       "390                                   CCCC           CCTC   \n",
       "392                                    CCC            CTC   \n",
       "406                                    AAG            AAA   \n",
       "394                                  ATACT          ATGTT   \n",
       "..                                     ...            ...   \n",
       "427                                TAGAATT        ATTCCCC   \n",
       "53                                     AAG            CCA   \n",
       "260                                    TGT            AAC   \n",
       "309                                    ACA            TTC   \n",
       "419                                 TAGACG         AGACAC   \n",
       "\n",
       "    Right_Flank_shortened  Hamming_Left  Hamming_Right  \n",
       "389                  CTCC      0.750000       0.500000  \n",
       "390                  TCCC      0.750000       0.500000  \n",
       "392                   TCC      0.666667       0.333333  \n",
       "406                   TTG      0.666667       0.000000  \n",
       "394                 AACCC      0.600000       0.200000  \n",
       "..                    ...           ...            ...  \n",
       "427               AATTCCC      0.000000       0.714286  \n",
       "53                    TTA      0.000000       0.333333  \n",
       "260                   ACG      0.000000       0.333333  \n",
       "309                   ACC      0.000000       0.333333  \n",
       "419                TAGACA      0.000000       0.000000  \n",
       "\n",
       "[49 rows x 5 columns]"
      ]
     },
     "execution_count": 784,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show rows where inserted bases are longer than 2.\n",
    "hamming_cutoff = 1\n",
    "table[\n",
    "\t(table['Inserted_bases_len'] > 2) &\n",
    "\t((table['Hamming_Left'].isnull()) | (table['Hamming_Left'] < hamming_cutoff)) &\n",
    "\t((table['Hamming_Right'].isnull()) | (table['Hamming_Right'] < hamming_cutoff))\n",
    "][['Left_Flank_flipped_shortened_unflipped', 'Inserted_bases', 'Right_Flank_shortened', 'Hamming_Left', 'Hamming_Right']].sort_values(by=['Hamming_Left', 'Hamming_Right'], ascending=[False, False])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "319a903c",
   "metadata": {},
   "source": [
    "### Summary table using Hamming distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 785,
   "id": "91cdbd05",
   "metadata": {
    "vscode": {
     "languageId": "ruby"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>1-bp</th>\n",
       "      <th>&gt;1-bp</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Species</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">mouse</th>\n",
       "      <th>Identical</th>\n",
       "      <td>24</td>\n",
       "      <td>8</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Different</th>\n",
       "      <td>18</td>\n",
       "      <td>8</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <td>42</td>\n",
       "      <td>16</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">macaque</th>\n",
       "      <th>Identical</th>\n",
       "      <td>155</td>\n",
       "      <td>54</td>\n",
       "      <td>209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Different</th>\n",
       "      <td>76</td>\n",
       "      <td>39</td>\n",
       "      <td>115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <td>231</td>\n",
       "      <td>93</td>\n",
       "      <td>324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">human</th>\n",
       "      <th>Identical</th>\n",
       "      <td>14</td>\n",
       "      <td>6</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Different</th>\n",
       "      <td>30</td>\n",
       "      <td>18</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <td>44</td>\n",
       "      <td>24</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   1-bp  >1-bp  Total\n",
       "Species                              \n",
       "mouse   Identical    24      8     32\n",
       "        Different    18      8     26\n",
       "        Total        42     16     58\n",
       "macaque Identical   155     54    209\n",
       "        Different    76     39    115\n",
       "        Total       231     93    324\n",
       "human   Identical    14      6     20\n",
       "        Different    30     18     48\n",
       "        Total        44     24     68"
      ]
     },
     "execution_count": 785,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def summarize_hamming(species, table=table, hamming_threshold=0.7):\n",
    "\t\"\"\"\n",
    "\tSummarize the hamming distance of insertions to their flanking sequences, excluding partial matches.\n",
    "\t\"\"\"\n",
    "\t# Filter table for the specified species\n",
    "\ttable = table.copy()[table['Species'] == species]\n",
    "\n",
    "\t# Create mask for hamming distance of 1 to either flank\n",
    "\tidentical_to_flank = (\n",
    "\t\t(table['Hamming_Left'] == 1) | (table['Hamming_Right'] == 1)\n",
    "\t)\n",
    "\n",
    "\t# Only count Identical and Different (exclude Partial match)\n",
    "\tsummary_hamming = pd.DataFrame({\n",
    "\t\t'1-bp': [\n",
    "\t\t\t((table['Inserted_bases_len'] == 1) & identical_to_flank).sum(),\n",
    "\t\t\t((table['Inserted_bases_len'] == 1) & ~identical_to_flank).sum(),\n",
    "\t\t\t(table['Inserted_bases_len'] == 1).sum()\n",
    "\t\t],\n",
    "\t\t'>1-bp': [\n",
    "\t\t\t((table['Inserted_bases_len'] > 1) & identical_to_flank).sum(),\n",
    "\t\t\t((table['Inserted_bases_len'] > 1) & ~identical_to_flank).sum(),\n",
    "\t\t\t(table['Inserted_bases_len'] > 1).sum()\n",
    "\t\t],\n",
    "\t\t'Total': [\n",
    "\t\t\tidentical_to_flank.sum(),\n",
    "\t\t\t(~identical_to_flank).sum(),\n",
    "\t\t\ttable.shape[0]\n",
    "\t\t],\n",
    "\t\t'Species': species\n",
    "\t}, index=['Identical', 'Different', 'Total'])\n",
    "\n",
    "\treturn summary_hamming\n",
    "\n",
    "# Run summarize_hamming for all species and concatenate the results into a single DataFrame\n",
    "summary_hamming_all = [\n",
    "\tsummarize_hamming(sp, table=table) for sp in species_list\n",
    "]\n",
    "summary_hamming_all_df = pd.concat(summary_hamming_all, keys=species_list, names=['Species']).drop(columns=['Species'])\n",
    "summary_hamming_all_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d086017",
   "metadata": {},
   "source": [
    "# Deletions:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c34d28f",
   "metadata": {},
   "source": [
    "## Parse FastA files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 786,
   "id": "7bee0d26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Deleted_bases_len</th>\n",
       "      <th>Left_Flank_flipped_shortened_unflipped</th>\n",
       "      <th>Deleted_bases</th>\n",
       "      <th>Right_Flank_deleted_shortened</th>\n",
       "      <th>Right_Flank_deleted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>TATATTCTCCAACAACAACGACAATCTAATTCCACTTATAGGCCTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>58</td>\n",
       "      <td>GAAACAGGATCAAACAACCCAACAGGATTAAACTCAGATGCAGATA...</td>\n",
       "      <td>ACCCCTACTATACAATCAAAGATATCCTAGGTATCCTAATCATATT...</td>\n",
       "      <td>AACCCTAGTATTATTTTTCCCAGACATACTAGGAGACCCAGACAAC...</td>\n",
       "      <td>AACCCTAGTATTATTTTTCCCAGACATACTAGGAGACCCAGACAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>CGAAAAAAAAAAATGGCGGTAGAAGTCTTAGTAGAGATTTCTCTAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>AAAAAAACCCACGATCAACTGAAGCAGCAACAAAATACTTCGTCAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>ACC</td>\n",
       "      <td>AAT</td>\n",
       "      <td>AAT</td>\n",
       "      <td>AATAATTGGAGGCTTTGGAAACTGACTTGTCCCACTAATAATCGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1092</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>CCCCCTCCCCATGCTTACAAGCAAGTACAGCAATCAACCCTCAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1093</th>\n",
       "      <td>2</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>TC</td>\n",
       "      <td>TCCCCATGCTTACAAGCAAGTACAGCAATCAACCCTCAACTATCAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1094</th>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>TCCCCATGCTTACAAGCAAGTACAGCAATCAACCCTCAACTATCAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1095</th>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>ACCCCTCACCCACTAGGATACCAACAAACCTACCCACCCTTAACAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1096</th>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>TCACCCACTAGGATACCAACAAACCTACCCACCCTTAACAGTACAT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1097 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Deleted_bases_len             Left_Flank_flipped_shortened_unflipped  \\\n",
       "0                     1                                                  A   \n",
       "1                    58  GAAACAGGATCAAACAACCCAACAGGATTAAACTCAGATGCAGATA...   \n",
       "2                     1                                                  G   \n",
       "3                     1                                                  C   \n",
       "4                     3                                                ACC   \n",
       "...                 ...                                                ...   \n",
       "1092                  1                                                  A   \n",
       "1093                  2                                                 CC   \n",
       "1094                  1                                                  C   \n",
       "1095                  1                                                  C   \n",
       "1096                  1                                                  C   \n",
       "\n",
       "                                          Deleted_bases  \\\n",
       "0                                                     T   \n",
       "1     ACCCCTACTATACAATCAAAGATATCCTAGGTATCCTAATCATATT...   \n",
       "2                                                     C   \n",
       "3                                                     A   \n",
       "4                                                   AAT   \n",
       "...                                                 ...   \n",
       "1092                                                  C   \n",
       "1093                                                 CT   \n",
       "1094                                                  T   \n",
       "1095                                                  A   \n",
       "1096                                                  T   \n",
       "\n",
       "                          Right_Flank_deleted_shortened  \\\n",
       "0                                                     T   \n",
       "1     AACCCTAGTATTATTTTTCCCAGACATACTAGGAGACCCAGACAAC...   \n",
       "2                                                     C   \n",
       "3                                                     A   \n",
       "4                                                   AAT   \n",
       "...                                                 ...   \n",
       "1092                                                  C   \n",
       "1093                                                 TC   \n",
       "1094                                                  T   \n",
       "1095                                                  A   \n",
       "1096                                                  T   \n",
       "\n",
       "                                    Right_Flank_deleted  \n",
       "0     TATATTCTCCAACAACAACGACAATCTAATTCCACTTATAGGCCTA...  \n",
       "1     AACCCTAGTATTATTTTTCCCAGACATACTAGGAGACCCAGACAAC...  \n",
       "2     CGAAAAAAAAAAATGGCGGTAGAAGTCTTAGTAGAGATTTCTCTAC...  \n",
       "3     AAAAAAACCCACGATCAACTGAAGCAGCAACAAAATACTTCGTCAC...  \n",
       "4     AATAATTGGAGGCTTTGGAAACTGACTTGTCCCACTAATAATCGGA...  \n",
       "...                                                 ...  \n",
       "1092  CCCCCTCCCCATGCTTACAAGCAAGTACAGCAATCAACCCTCAACT...  \n",
       "1093  TCCCCATGCTTACAAGCAAGTACAGCAATCAACCCTCAACTATCAC...  \n",
       "1094  TCCCCATGCTTACAAGCAAGTACAGCAATCAACCCTCAACTATCAC...  \n",
       "1095  ACCCCTCACCCACTAGGATACCAACAAACCTACCCACCCTTAACAG...  \n",
       "1096  TCACCCACTAGGATACCAACAAACCTACCCACCCTTAACAGTACAT...  \n",
       "\n",
       "[1097 rows x 5 columns]"
      ]
     },
     "execution_count": 786,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_rows', 20)\n",
    "\n",
    "mutation_type = 'del'  # 'ins' or 'del'\n",
    "\n",
    "all_tables = []\n",
    "\n",
    "for species in species_list:\n",
    "\tfasta_file = f\"{species}.indels.{mutation_type}.slop{slop_size}.fa\"\n",
    "\tfasta_dict = {record.id: str(record.seq) for record in SeqIO.parse(fasta_file, \"fasta\")}\n",
    "\n",
    "\tsplit_fasta_dict = {}\n",
    "\tfor key, seq in fasta_dict.items():\n",
    "\t\tmid = len(seq) // 2\n",
    "\t\tsplit_fasta_dict[key] = (seq[:mid+1], seq[mid+1:])\n",
    "\n",
    "\ttable = pd.DataFrame([\n",
    "\t\t{'Fasta_ID': k, 'Left_Flank': v[0], 'Right_Flank': v[1]}\n",
    "\t\tfor k, v in split_fasta_dict.items()\n",
    "\t])\n",
    "\n",
    "\t# Extract Mutation_ID from Fasta_ID.\n",
    "\ttable['Mutation_ID'] = table['Fasta_ID'].str.split(':').str[0].str.split('_').str[1:4].str.join('_')\n",
    "\n",
    "\ttable['Left_Flank'] = table['Left_Flank'].str.upper()\n",
    "\ttable['Right_Flank'] = table['Right_Flank'].str.upper()\n",
    "\ttable['Left_Flank_flipped'] = table['Left_Flank'].apply(lambda x: x[::-1])\n",
    "\ttable[['Species', 'Position', 'Ref', 'Alt', 'Type', 'Region']] = table['Fasta_ID'].str.extract(r'([^_]+)_([^_]+)_([^_]+)_([^_]+)_([^:]+)::(.+)')\n",
    "\tif mutation_type == 'del':\n",
    "\t\ttable['Deleted_bases'] = table['Ref'].str[1:]\n",
    "\telse:\n",
    "\t\traise ValueError(\"Currently only 'del' mutation_type is supported.\")\n",
    "\ttable['Deleted_bases_len'] = table['Deleted_bases'].str.len()\n",
    "\ttable['Right_Flank_deleted'] = table.apply(\n",
    "\t\tlambda row: row['Right_Flank'][row['Deleted_bases_len']:] if row['Deleted_bases_len'] > 0 else row['Right_Flank'],\n",
    "\t\taxis=1\n",
    "\t)\n",
    "\n",
    "\t# Consider only the flank that is the same size as the deletion.\n",
    "\ttable['Left_Flank_flipped_shortened'] = table.apply(lambda row: \n",
    "\t\t\t\t\t\t\t\t\trow['Left_Flank_flipped'][:row['Deleted_bases_len']], axis=1)\n",
    "\ttable['Right_Flank_deleted_shortened'] = table.apply(lambda row: \n",
    "\t\t\t\t\t\t\t\t\trow['Right_Flank_deleted'][:row['Deleted_bases_len']], axis=1)\n",
    "\ttable['Right_Flank_deleted_shortened_flipped'] = table['Right_Flank_deleted_shortened'].apply(lambda x: x[::-1])\n",
    "\ttable['Left_Flank_flipped_shortened_unflipped'] = table['Left_Flank_flipped_shortened'].apply(lambda x: x[::-1])\n",
    "\n",
    "\tall_tables.append(table)\n",
    "\n",
    "# Concatenate all tables into one DataFrame\n",
    "table = pd.concat(all_tables, ignore_index=True)\n",
    "table[['Deleted_bases_len', 'Left_Flank_flipped_shortened_unflipped', 'Deleted_bases', 'Right_Flank_deleted_shortened', 'Right_Flank_deleted']]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e0fcb91",
   "metadata": {},
   "source": [
    "### Hamming distance between inserted bases and flanks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 787,
   "id": "5f493d49",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Left_Flank_flipped_shortened_unflipped</th>\n",
       "      <th>Deleted_bases</th>\n",
       "      <th>Right_Flank_deleted_shortened</th>\n",
       "      <th>Hamming_Left</th>\n",
       "      <th>Hamming_Right</th>\n",
       "      <th>Identical_to_flank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GAAACAGGATCAAACAACCCAACAGGATTAAACTCAGATGCAGATA...</td>\n",
       "      <td>ACCCCTACTATACAATCAAAGATATCCTAGGTATCCTAATCATATT...</td>\n",
       "      <td>AACCCTAGTATTATTTTTCCCAGACATACTAGGAGACCCAGACAAC...</td>\n",
       "      <td>0.241379</td>\n",
       "      <td>0.293103</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACC</td>\n",
       "      <td>AAT</td>\n",
       "      <td>AAT</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1092</th>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1093</th>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>TC</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1094</th>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1095</th>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1096</th>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1097 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Left_Flank_flipped_shortened_unflipped  \\\n",
       "0                                                     A   \n",
       "1     GAAACAGGATCAAACAACCCAACAGGATTAAACTCAGATGCAGATA...   \n",
       "2                                                     G   \n",
       "3                                                     C   \n",
       "4                                                   ACC   \n",
       "...                                                 ...   \n",
       "1092                                                  A   \n",
       "1093                                                 CC   \n",
       "1094                                                  C   \n",
       "1095                                                  C   \n",
       "1096                                                  C   \n",
       "\n",
       "                                          Deleted_bases  \\\n",
       "0                                                     T   \n",
       "1     ACCCCTACTATACAATCAAAGATATCCTAGGTATCCTAATCATATT...   \n",
       "2                                                     C   \n",
       "3                                                     A   \n",
       "4                                                   AAT   \n",
       "...                                                 ...   \n",
       "1092                                                  C   \n",
       "1093                                                 CT   \n",
       "1094                                                  T   \n",
       "1095                                                  A   \n",
       "1096                                                  T   \n",
       "\n",
       "                          Right_Flank_deleted_shortened  Hamming_Left  \\\n",
       "0                                                     T      0.000000   \n",
       "1     AACCCTAGTATTATTTTTCCCAGACATACTAGGAGACCCAGACAAC...      0.241379   \n",
       "2                                                     C      0.000000   \n",
       "3                                                     A      0.000000   \n",
       "4                                                   AAT      0.333333   \n",
       "...                                                 ...           ...   \n",
       "1092                                                  C      0.000000   \n",
       "1093                                                 TC      0.500000   \n",
       "1094                                                  T      0.000000   \n",
       "1095                                                  A      0.000000   \n",
       "1096                                                  T      0.000000   \n",
       "\n",
       "      Hamming_Right  Identical_to_flank  \n",
       "0          1.000000                True  \n",
       "1          0.293103               False  \n",
       "2          1.000000                True  \n",
       "3          1.000000                True  \n",
       "4          1.000000                True  \n",
       "...             ...                 ...  \n",
       "1092       1.000000                True  \n",
       "1093       0.000000               False  \n",
       "1094       1.000000                True  \n",
       "1095       1.000000                True  \n",
       "1096       1.000000                True  \n",
       "\n",
       "[1097 rows x 6 columns]"
      ]
     },
     "execution_count": 787,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from difflib import SequenceMatcher\n",
    "\n",
    "def hamming_similarity(s1, s2):\n",
    "\t\"\"\"Return the proportion of matching characters (Hamming similarity) between two strings of equal length.\"\"\"\n",
    "\tif len(s1) != len(s2) or len(s1) == 0:\n",
    "\t\treturn None\n",
    "\treturn sum(a == b for a, b in zip(s1, s2)) / len(s1)\n",
    "\n",
    "# Compute Hamming similarity for each row between Deleted_bases and Left_Flank_flipped_shortened, Right_Flank_deleted_shortened.\n",
    "table['Hamming_Left'] = table.apply(\n",
    "\tlambda row: hamming_similarity(row['Deleted_bases'], row['Left_Flank_flipped_shortened_unflipped']), axis=1\n",
    ")\n",
    "table['Hamming_Right'] = table.apply(\n",
    "\tlambda row: hamming_similarity(row['Deleted_bases'], row['Right_Flank_deleted_shortened']), axis=1\n",
    ")\n",
    "\n",
    "# Identify rows where the deleted bases are identical to either flank.\n",
    "table['Identical_to_flank'] = (table['Hamming_Left'] == 1) | (table['Hamming_Right'] == 1)\n",
    "\n",
    "# Create a copy of the table for deletions.\n",
    "table_deletions = table.copy()\n",
    "\n",
    "table[['Left_Flank_flipped_shortened_unflipped', 'Deleted_bases', 'Right_Flank_deleted_shortened', 'Hamming_Left', 'Hamming_Right', 'Identical_to_flank']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5731cd14",
   "metadata": {},
   "source": [
    "### Summary table using Hamming distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bfedfcf",
   "metadata": {
    "vscode": {
     "languageId": "ruby"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>1-bp</th>\n",
       "      <th>&gt;1-bp</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Species</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">mouse</th>\n",
       "      <th>Identical</th>\n",
       "      <td>96</td>\n",
       "      <td>21</td>\n",
       "      <td>117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Different</th>\n",
       "      <td>35</td>\n",
       "      <td>50</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <td>131</td>\n",
       "      <td>71</td>\n",
       "      <td>202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">macaque</th>\n",
       "      <th>Identical</th>\n",
       "      <td>318</td>\n",
       "      <td>76</td>\n",
       "      <td>394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Different</th>\n",
       "      <td>118</td>\n",
       "      <td>304</td>\n",
       "      <td>422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <td>436</td>\n",
       "      <td>380</td>\n",
       "      <td>816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">human</th>\n",
       "      <th>Identical</th>\n",
       "      <td>39</td>\n",
       "      <td>3</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Different</th>\n",
       "      <td>7</td>\n",
       "      <td>30</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <td>46</td>\n",
       "      <td>33</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   1-bp  >1-bp  Total\n",
       "Species                              \n",
       "mouse   Identical    96     21    117\n",
       "        Different    35     50     85\n",
       "        Total       131     71    202\n",
       "macaque Identical   318     76    394\n",
       "        Different   118    304    422\n",
       "        Total       436    380    816\n",
       "human   Identical    39      3     42\n",
       "        Different     7     30     37\n",
       "        Total        46     33     79"
      ]
     },
     "execution_count": 788,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def summarize_hamming(species, table=table, hamming_threshold=0.7):\n",
    "\t\"\"\"\n",
    "\tSummarize the hamming distance of deletions to their flanking sequences, excluding partial matches.\n",
    "\t\"\"\"\n",
    "\t# Filter table for the specified species\n",
    "\ttable = table.copy()[table['Species'] == species]\n",
    "\n",
    "\t# Create mask for hamming distance of 1 to either flank\n",
    "\tidentical_to_flank = (\n",
    "\t\t(table['Hamming_Left'] == 1) | (table['Hamming_Right'] == 1)\n",
    "\t)\n",
    "\n",
    "\t# Create mask for hamming distance <1 but >hamming_threshold to either flank (partial match)\n",
    "\tpartial_match_to_flank = (\n",
    "\t\t(((table['Hamming_Left'] < 1) & (table['Hamming_Left'] > hamming_threshold)) |\n",
    "\t\t  ((table['Hamming_Right'] < 1) & (table['Hamming_Right'] > hamming_threshold)))\n",
    "\t)\n",
    "\n",
    "\t# Only count Identical and Different (exclude Partial match)\n",
    "\tsummary_hamming = pd.DataFrame({\n",
    "\t\t'1-bp': [\n",
    "\t\t\t((table['Deleted_bases_len'] == 1) & identical_to_flank).sum(),\n",
    "\t\t\t((table['Deleted_bases_len'] == 1) & ~identical_to_flank).sum(),\n",
    "\t\t\t(table['Deleted_bases_len'] == 1).sum()\n",
    "\t\t],\n",
    "\t\t'>1-bp': [\n",
    "\t\t\t((table['Deleted_bases_len'] > 1) & identical_to_flank).sum(),\n",
    "\t\t\t((table['Deleted_bases_len'] > 1) & ~identical_to_flank).sum(),\n",
    "\t\t\t(table['Deleted_bases_len'] > 1).sum()\n",
    "\t\t],\n",
    "\t\t'Total': [\n",
    "\t\t\tidentical_to_flank.sum(),\n",
    "\t\t\t(~identical_to_flank).sum(),\n",
    "\t\t\ttable.shape[0]\n",
    "\t\t],\n",
    "\t\t'Species': species\n",
    "\t}, index=['Identical', 'Different', 'Total'])\n",
    "\n",
    "\treturn summary_hamming\n",
    "\n",
    "# Run summarize_hamming for all species and concatenate the results into a single DataFrame\n",
    "summary_hamming_all = [\n",
    "\tsummarize_hamming(sp, table=table) for sp in species_list\n",
    "]\n",
    "summary_hamming_all_df = pd.concat(summary_hamming_all, keys=species_list, names=['Species']).drop(columns=['Species'])\n",
    "summary_hamming_all_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6aee24b9",
   "metadata": {},
   "source": [
    "# Indels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 789,
   "id": "bd5f44ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Mutation_ID</th>\n",
       "      <th>Identical_to_flank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mouse</td>\n",
       "      <td>12343_AT_A</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mouse</td>\n",
       "      <td>14806_CACCCCTACTATACAATCAAAGATATCCTAGGTATCCTAA...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mouse</td>\n",
       "      <td>5168_GC_G</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mouse</td>\n",
       "      <td>4051_CA_C</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mouse</td>\n",
       "      <td>5542_CAAT_C</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>445</th>\n",
       "      <td>human</td>\n",
       "      <td>16182_A_AC</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>446</th>\n",
       "      <td>human</td>\n",
       "      <td>16188_T_TC</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447</th>\n",
       "      <td>human</td>\n",
       "      <td>16188_T_TCC</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>448</th>\n",
       "      <td>human</td>\n",
       "      <td>16356_T_TCTCGTCCCCATGGATGACCCCC</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>449</th>\n",
       "      <td>human</td>\n",
       "      <td>16373_A_AC</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1547 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Species                                        Mutation_ID  \\\n",
       "0     mouse                                         12343_AT_A   \n",
       "1     mouse  14806_CACCCCTACTATACAATCAAAGATATCCTAGGTATCCTAA...   \n",
       "2     mouse                                          5168_GC_G   \n",
       "3     mouse                                          4051_CA_C   \n",
       "4     mouse                                        5542_CAAT_C   \n",
       "..      ...                                                ...   \n",
       "445   human                                         16182_A_AC   \n",
       "446   human                                         16188_T_TC   \n",
       "447   human                                        16188_T_TCC   \n",
       "448   human                    16356_T_TCTCGTCCCCATGGATGACCCCC   \n",
       "449   human                                         16373_A_AC   \n",
       "\n",
       "     Identical_to_flank  \n",
       "0                  True  \n",
       "1                 False  \n",
       "2                  True  \n",
       "3                  True  \n",
       "4                  True  \n",
       "..                  ...  \n",
       "445               False  \n",
       "446                True  \n",
       "447                True  \n",
       "448               False  \n",
       "449               False  \n",
       "\n",
       "[1547 rows x 3 columns]"
      ]
     },
     "execution_count": 789,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table_indels = pd.concat([table_deletions[['Species','Mutation_ID','Identical_to_flank']],table_insertions[['Species','Mutation_ID','Identical_to_flank']]])\n",
    "table_indels"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "320cc0d6",
   "metadata": {},
   "source": [
    "# Microsatellites/STRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 790,
   "id": "49ac54b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Mutation_ID</th>\n",
       "      <th>Change_in_length</th>\n",
       "      <th>Within_Repeat</th>\n",
       "      <th>Identical_to_flank</th>\n",
       "      <th>Mutation_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mouse</td>\n",
       "      <td>12343_AT_A</td>\n",
       "      <td>-1</td>\n",
       "      <td>STR</td>\n",
       "      <td>True</td>\n",
       "      <td>Deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mouse</td>\n",
       "      <td>14806_CACCCCTACTATACAATCAAAGATATCCTAGGTATCCTAA...</td>\n",
       "      <td>-58</td>\n",
       "      <td>Non-STR</td>\n",
       "      <td>False</td>\n",
       "      <td>Deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mouse</td>\n",
       "      <td>5168_GC_G</td>\n",
       "      <td>-1</td>\n",
       "      <td>Non-STR</td>\n",
       "      <td>True</td>\n",
       "      <td>Deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mouse</td>\n",
       "      <td>5168_GC_G</td>\n",
       "      <td>-1</td>\n",
       "      <td>Non-STR</td>\n",
       "      <td>True</td>\n",
       "      <td>Deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mouse</td>\n",
       "      <td>5168_GC_G</td>\n",
       "      <td>-1</td>\n",
       "      <td>Non-STR</td>\n",
       "      <td>True</td>\n",
       "      <td>Deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2712</th>\n",
       "      <td>human</td>\n",
       "      <td>16188_T_TC</td>\n",
       "      <td>1</td>\n",
       "      <td>STR</td>\n",
       "      <td>True</td>\n",
       "      <td>Insertion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2713</th>\n",
       "      <td>human</td>\n",
       "      <td>16188_T_TCC</td>\n",
       "      <td>2</td>\n",
       "      <td>STR</td>\n",
       "      <td>True</td>\n",
       "      <td>Insertion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2714</th>\n",
       "      <td>human</td>\n",
       "      <td>16356_T_TCTCGTCCCCATGGATGACCCCC</td>\n",
       "      <td>22</td>\n",
       "      <td>Non-STR</td>\n",
       "      <td>False</td>\n",
       "      <td>Insertion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2715</th>\n",
       "      <td>human</td>\n",
       "      <td>16373_A_AC</td>\n",
       "      <td>1</td>\n",
       "      <td>Non-STR</td>\n",
       "      <td>False</td>\n",
       "      <td>Insertion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2716</th>\n",
       "      <td>human</td>\n",
       "      <td>16373_A_AC</td>\n",
       "      <td>1</td>\n",
       "      <td>Non-STR</td>\n",
       "      <td>False</td>\n",
       "      <td>Insertion</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2717 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Species                                        Mutation_ID  \\\n",
       "0      mouse                                         12343_AT_A   \n",
       "1      mouse  14806_CACCCCTACTATACAATCAAAGATATCCTAGGTATCCTAA...   \n",
       "2      mouse                                          5168_GC_G   \n",
       "3      mouse                                          5168_GC_G   \n",
       "4      mouse                                          5168_GC_G   \n",
       "...      ...                                                ...   \n",
       "2712   human                                         16188_T_TC   \n",
       "2713   human                                        16188_T_TCC   \n",
       "2714   human                    16356_T_TCTCGTCCCCATGGATGACCCCC   \n",
       "2715   human                                         16373_A_AC   \n",
       "2716   human                                         16373_A_AC   \n",
       "\n",
       "      Change_in_length Within_Repeat  Identical_to_flank Mutation_type  \n",
       "0                   -1           STR                True      Deletion  \n",
       "1                  -58       Non-STR               False      Deletion  \n",
       "2                   -1       Non-STR                True      Deletion  \n",
       "3                   -1       Non-STR                True      Deletion  \n",
       "4                   -1       Non-STR                True      Deletion  \n",
       "...                ...           ...                 ...           ...  \n",
       "2712                 1           STR                True     Insertion  \n",
       "2713                 2           STR                True     Insertion  \n",
       "2714                22       Non-STR               False     Insertion  \n",
       "2715                 1       Non-STR               False     Insertion  \n",
       "2716                 1       Non-STR               False     Insertion  \n",
       "\n",
       "[2717 rows x 6 columns]"
      ]
     },
     "execution_count": 790,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "repeats = pd.read_table(\"indels.denovo_chrM_microsats.tab\")\n",
    "\n",
    "# Merge with the main indels table.\n",
    "table_indels_repeats = pd.merge(repeats[['Species','Mutation_ID','Change_in_length','Within_Repeat']],\n",
    "\t\t table_indels[['Species','Mutation_ID','Identical_to_flank']], how='right')#.drop_duplicates()\n",
    "table_indels_repeats['Mutation_type'] = table_indels_repeats['Change_in_length'].apply(lambda x: 'Insertion' if x > 0 else 'Deletion')\t\n",
    "table_indels_repeats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17c979f9",
   "metadata": {
    "vscode": {
     "languageId": "ruby"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Within_Repeat</th>\n",
       "      <th>Species</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Identical to Flank</th>\n",
       "      <th>Non-STR</th>\n",
       "      <th>STR</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mouse</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>Not identical</td>\n",
       "      <td>63</td>\n",
       "      <td>39</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mouse</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>Identical</td>\n",
       "      <td>56</td>\n",
       "      <td>154</td>\n",
       "      <td>210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mouse</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>Not identical</td>\n",
       "      <td>16</td>\n",
       "      <td>10</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mouse</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>Identical</td>\n",
       "      <td>3</td>\n",
       "      <td>86</td>\n",
       "      <td>89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>macaque</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>Not identical</td>\n",
       "      <td>298</td>\n",
       "      <td>337</td>\n",
       "      <td>635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>macaque</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>Identical</td>\n",
       "      <td>212</td>\n",
       "      <td>585</td>\n",
       "      <td>797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>macaque</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>Not identical</td>\n",
       "      <td>80</td>\n",
       "      <td>83</td>\n",
       "      <td>163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>macaque</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>Identical</td>\n",
       "      <td>92</td>\n",
       "      <td>420</td>\n",
       "      <td>512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>human</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>Not identical</td>\n",
       "      <td>29</td>\n",
       "      <td>12</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>human</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>Identical</td>\n",
       "      <td>25</td>\n",
       "      <td>24</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>human</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>Not identical</td>\n",
       "      <td>37</td>\n",
       "      <td>26</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>human</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>Identical</td>\n",
       "      <td>12</td>\n",
       "      <td>18</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Within_Repeat  Species Mutation Type Identical to Flank  Non-STR  STR  Total\n",
       "0                mouse      Deletion      Not identical       63   39    102\n",
       "1                mouse      Deletion          Identical       56  154    210\n",
       "2                mouse     Insertion      Not identical       16   10     26\n",
       "3                mouse     Insertion          Identical        3   86     89\n",
       "4              macaque      Deletion      Not identical      298  337    635\n",
       "5              macaque      Deletion          Identical      212  585    797\n",
       "6              macaque     Insertion      Not identical       80   83    163\n",
       "7              macaque     Insertion          Identical       92  420    512\n",
       "8                human      Deletion      Not identical       29   12     41\n",
       "9                human      Deletion          Identical       25   24     49\n",
       "10               human     Insertion      Not identical       37   26     63\n",
       "11               human     Insertion          Identical       12   18     30"
      ]
     },
     "execution_count": 791,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Summarize counts by Species, Within_Repeat, and Identical_to_flank\n",
    "summary_repeat_flank = (\n",
    "\ttable_indels_repeats\n",
    "\t.groupby(['Species', 'Mutation_type', 'Identical_to_flank', 'Within_Repeat'])\n",
    "\t.size()\n",
    "\t.unstack(fill_value=0)\n",
    ")\n",
    "# Add a Total column.\n",
    "summary_repeat_flank['Total'] = summary_repeat_flank.sum(axis=1)\n",
    "\n",
    "# Reorder Species index to mouse, macaque, human\n",
    "species_order = ['mouse', 'macaque', 'human']\n",
    "summary_repeat_flank = summary_repeat_flank.reindex(species_order, level='Species')\n",
    "summary_repeat_flank = summary_repeat_flank.rename_axis(index={'Species': 'Species', 'Mutation_type': 'Mutation Type', 'Identical_to_flank': 'Identical to Flank', 'Within_Repeat': 'Within Repeat'})\n",
    "summary_repeat_flank = summary_repeat_flank.rename({True: 'Identical', False: 'Not identical'})\n",
    "summary_repeat_flank = summary_repeat_flank.reset_index()\n",
    "summary_repeat_flank.reset_index().rename(columns={\n",
    "\t'Species': 'Species',\n",
    "\t'Mutation_type': 'Mutation Type',\n",
    "\t'Identical_to_flank': 'Identical to Flank',\n",
    "\t'Within_Repeat': 'Within Repeat'\n",
    "})\n",
    "summary_repeat_flank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d374863c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96a3c336",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
